{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Pandas\n",
    "In this notebook, we'll encounter a very useful package for scientific computing in Python: Pandas. We can think of Pandas as \"numpy with labels\". This package is especially useful for data science and biology for a few reasons:\n",
    "* Great for real-world, heterogeneous data\n",
    "* Similar to Excel spreadsheets (but way faster!)\n",
    "* Smartly deals with missing data\n",
    "\n",
    "### At the end of this notebook, you'll be able to:\n",
    "* Create & manipulate Pandas dataframes\n",
    "* Load a tabular file of genetic data into a Pandas dataframe\n",
    "* Index and subset Pandas dataframes\n",
    "* Generate descriptive statistics for Pandas dataframes\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Importing packages\n",
    "\n",
    "Before we can pandas, we need to import it. The convention is to import `pandas` as `pd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable   Type      Data/Info\n",
      "------------------------------\n",
      "pd         module    <module 'pandas' from '/o<...>ages/pandas/__init__.py'>\n"
     ]
    }
   ],
   "source": [
    "# Import packages\n",
    "import pandas as pd\n",
    "\n",
    "# Use whos 'magic command' to see available modules\n",
    "%whos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Create and Manipulate Dataframes \n",
    "The two data structures of Pandas are the `Series` and the `DataFrame`. A `Series` is a one-dimensional onject similar to a list. A `DataFrame` can be thought of as a two-dimensional numpy array or a collection of `Series` objects. Series and dataframes can contain multiple different data types such as integers, strings, and floats, similar to an Excel spreadsheet. Pandas also supports `string` lables unlike numpy arrays which only have numeric labels for their rows and columns. For a more in depth explanation, please visit the [Introduction to Data Structures](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html) section in the Pandas User Guide. \n",
    "\n",
    "You can create a Pandas dataframe by inputting dictionaries into the Pandas function `pd.DataFrame()`, by reading files, or through functions built into the Pandas package. The function [`pd.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) reads a comma- or tab-separated file and returns it as a `dataframe`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a data from a dictionary\n",
    "\n",
    "Below, we'll create a dataframe from a dictionary. Since this is quite a bit to fit into one cell, we're also using a trick where we use a pair of open parentheses to tell Python to *ignore* the line breaks here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FIRST_NAME</th>\n",
       "      <th>LAST_NAME</th>\n",
       "      <th>YEAR</th>\n",
       "      <th>HOME_STATE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>CALC_101_FINAL</th>\n",
       "      <th>ENGLISH_101_FINAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Daniel</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>18</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ben</td>\n",
       "      <td>Leibstrom</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>19</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kavita</td>\n",
       "      <td>Kanabar</td>\n",
       "      <td>1</td>\n",
       "      <td>PA</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linda</td>\n",
       "      <td>Thiel</td>\n",
       "      <td>4</td>\n",
       "      <td>CA</td>\n",
       "      <td>22</td>\n",
       "      <td>60.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Omar</td>\n",
       "      <td>Reichel</td>\n",
       "      <td>2</td>\n",
       "      <td>OK</td>\n",
       "      <td>21</td>\n",
       "      <td>70.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Jane</td>\n",
       "      <td>OConner</td>\n",
       "      <td>2</td>\n",
       "      <td>HI</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Felicia</td>\n",
       "      <td>Rao</td>\n",
       "      <td>3</td>\n",
       "      <td>NY</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rachel</td>\n",
       "      <td>Crock</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bob</td>\n",
       "      <td>McDonald</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>18</td>\n",
       "      <td>98.0</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  FIRST_NAME  LAST_NAME  YEAR HOME_STATE  AGE  CALC_101_FINAL  \\\n",
       "0     Daniel      Smith     1         NY   18            90.0   \n",
       "1        Ben  Leibstrom     1         NY   19            80.0   \n",
       "2     Kavita    Kanabar     1         PA   19             NaN   \n",
       "3      Linda      Thiel     4         CA   22            60.0   \n",
       "4       Omar    Reichel     2         OK   21            70.0   \n",
       "5       Jane    OConner     2         HI   19             NaN   \n",
       "6    Felicia        Rao     3         NY   20             NaN   \n",
       "7     Rachel      Crock     1         FL   17             NaN   \n",
       "8        Bob   McDonald     1         FL   18            98.0   \n",
       "\n",
       "   ENGLISH_101_FINAL  \n",
       "0               80.0  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3               40.0  \n",
       "4               50.0  \n",
       "5                NaN  \n",
       "6                NaN  \n",
       "7               60.0  \n",
       "8               65.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the dictionary\n",
    "students = (\n",
    "  {'FIRST_NAME': ['Daniel', 'Ben', 'Kavita', 'Linda',\n",
    "                  'Omar','Jane', 'Felicia', 'Rachel',\n",
    "                  'Bob'],\n",
    "   'LAST_NAME': ['Smith', 'Leibstrom', 'Kanabar', 'Thiel',\n",
    "                 'Reichel', 'OConner', 'Rao', 'Crock',\n",
    "                 'McDonald'],\n",
    "   'YEAR': [1, 1, 1, 4, 2, 2, 3, 1, 1],\n",
    "   'HOME_STATE': ['NY', 'NY', 'PA', 'CA', 'OK', 'HI',\n",
    "                  'NY','FL', 'FL'],\n",
    "   'AGE': [18, 19, 19, 22, 21, 19, 20, 17, 18],\n",
    "   'CALC_101_FINAL': [90, 80, None, 60, 70, None, None,\n",
    "                      None, 98],\n",
    "   'ENGLISH_101_FINAL': [80, None, None, 40, 50, None,\n",
    "                         None, 60, 65]} )\n",
    "\n",
    "# Create the dataframe from our dictionary\n",
    "students_df = pd.DataFrame(students)\n",
    "students_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can access a list of all of the students' last names using the following notation: `df['column_name']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        Smith\n",
       "1    Leibstrom\n",
       "2      Kanabar\n",
       "3        Thiel\n",
       "4      Reichel\n",
       "5      OConner\n",
       "6          Rao\n",
       "7        Crock\n",
       "8     McDonald\n",
       "Name: LAST_NAME, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Access LAST_NAME column\n",
    "lastnames = students_df['LAST_NAME']\n",
    "lastnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One column of the dataframe is a **series**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.series.Series"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check type of column\n",
    "type(lastnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading from a file example\n",
    "Below we will create a dataframe by reading the file `brainarea_vs_genes_exp_w_reannotations.tsv` which contains information on gene expression accross multiple brain areas. we'll use the `read_csv` function to import our gene expression data as a DataFrame.\n",
    "\n",
    "<font color='red'><b>Note</b>:</font> This file is *not* located in your DataHub. It is available on Canvas and needs to be uploaded into your Data folder on DataHub.\n",
    "\n",
    ">**About this dataset:**\n",
    "This dataset was created by Derek Howard and Abigail Mayes for the purpose of accelerating advances in data mining of open brain transcriptome data for polygenetic brain disorders. The data comes from normalized microarray datasets of gene expression from 6 adult human brains that was released by the Allen Brain Institute and then processed into the dataframe we will see below. For more information on this dataset please visit the <a href = \"https://github.com/derekhoward/HBAsets\"> HBAsets repository</a>. \n",
    "\n",
    "Below, we'll use the `read_csv()` function to import our gene expression data as a **data frame**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "      <td>0.826986</td>\n",
       "      <td>0.948039</td>\n",
       "      <td>0.935427</td>\n",
       "      <td>1.120774</td>\n",
       "      <td>-1.018554</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>-0.104286</td>\n",
       "      <td>-0.910245</td>\n",
       "      <td>1.039610</td>\n",
       "      <td>-0.155167</td>\n",
       "      <td>-0.444398</td>\n",
       "      <td>-0.901361</td>\n",
       "      <td>-0.236790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "      <td>0.362799</td>\n",
       "      <td>0.353296</td>\n",
       "      <td>0.422766</td>\n",
       "      <td>0.346853</td>\n",
       "      <td>-0.812015</td>\n",
       "      <td>...</td>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.901213</td>\n",
       "      <td>-1.522431</td>\n",
       "      <td>0.598719</td>\n",
       "      <td>-1.709745</td>\n",
       "      <td>-0.054156</td>\n",
       "      <td>-1.695843</td>\n",
       "      <td>-1.155961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "      <td>0.507916</td>\n",
       "      <td>0.577696</td>\n",
       "      <td>0.647671</td>\n",
       "      <td>0.306824</td>\n",
       "      <td>0.089958</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.465863</td>\n",
       "      <td>-0.136936</td>\n",
       "      <td>1.229487</td>\n",
       "      <td>-0.110680</td>\n",
       "      <td>-0.118175</td>\n",
       "      <td>-0.139776</td>\n",
       "      <td>0.123829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "      <td>-1.687391</td>\n",
       "      <td>-1.756847</td>\n",
       "      <td>-1.640242</td>\n",
       "      <td>-1.733110</td>\n",
       "      <td>-0.091695</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.183630</td>\n",
       "      <td>0.948098</td>\n",
       "      <td>-0.977692</td>\n",
       "      <td>0.911896</td>\n",
       "      <td>-0.499357</td>\n",
       "      <td>1.469386</td>\n",
       "      <td>0.557998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "      <td>-0.289888</td>\n",
       "      <td>-0.407026</td>\n",
       "      <td>-0.358798</td>\n",
       "      <td>-0.589988</td>\n",
       "      <td>0.944684</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>-0.282673</td>\n",
       "      <td>1.396365</td>\n",
       "      <td>0.945043</td>\n",
       "      <td>0.158202</td>\n",
       "      <td>0.572771</td>\n",
       "      <td>0.073088</td>\n",
       "      <td>-0.886780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A3GALT2</td>\n",
       "      <td>0.435751</td>\n",
       "      <td>1.125971</td>\n",
       "      <td>1.479726</td>\n",
       "      <td>0.010520</td>\n",
       "      <td>0.877256</td>\n",
       "      <td>0.617134</td>\n",
       "      <td>0.797456</td>\n",
       "      <td>0.724969</td>\n",
       "      <td>0.692655</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.339939</td>\n",
       "      <td>-0.628313</td>\n",
       "      <td>-0.246285</td>\n",
       "      <td>0.049359</td>\n",
       "      <td>0.223667</td>\n",
       "      <td>1.164986</td>\n",
       "      <td>-0.102258</td>\n",
       "      <td>-0.849872</td>\n",
       "      <td>-0.354230</td>\n",
       "      <td>-0.421125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>1.277084</td>\n",
       "      <td>0.710266</td>\n",
       "      <td>0.272435</td>\n",
       "      <td>0.452544</td>\n",
       "      <td>0.127867</td>\n",
       "      <td>-0.423357</td>\n",
       "      <td>0.222945</td>\n",
       "      <td>-0.453995</td>\n",
       "      <td>-0.919359</td>\n",
       "      <td>...</td>\n",
       "      <td>0.100831</td>\n",
       "      <td>0.322003</td>\n",
       "      <td>0.622643</td>\n",
       "      <td>0.698657</td>\n",
       "      <td>0.060055</td>\n",
       "      <td>-1.546518</td>\n",
       "      <td>-0.578289</td>\n",
       "      <td>-1.117115</td>\n",
       "      <td>-0.442883</td>\n",
       "      <td>-0.352982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A4GNT</td>\n",
       "      <td>-0.330903</td>\n",
       "      <td>-0.477462</td>\n",
       "      <td>0.267584</td>\n",
       "      <td>-0.426862</td>\n",
       "      <td>0.223556</td>\n",
       "      <td>0.260993</td>\n",
       "      <td>0.632328</td>\n",
       "      <td>0.097173</td>\n",
       "      <td>-0.173487</td>\n",
       "      <td>...</td>\n",
       "      <td>0.555680</td>\n",
       "      <td>-0.462533</td>\n",
       "      <td>0.752745</td>\n",
       "      <td>0.648017</td>\n",
       "      <td>1.639433</td>\n",
       "      <td>0.353052</td>\n",
       "      <td>0.249200</td>\n",
       "      <td>0.934437</td>\n",
       "      <td>-0.165556</td>\n",
       "      <td>-0.715761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>-1.257280</td>\n",
       "      <td>-0.888406</td>\n",
       "      <td>-1.124108</td>\n",
       "      <td>-1.118677</td>\n",
       "      <td>1.234268</td>\n",
       "      <td>1.421483</td>\n",
       "      <td>1.429758</td>\n",
       "      <td>1.603573</td>\n",
       "      <td>-0.827011</td>\n",
       "      <td>...</td>\n",
       "      <td>0.073441</td>\n",
       "      <td>-0.372438</td>\n",
       "      <td>-0.564808</td>\n",
       "      <td>-0.011762</td>\n",
       "      <td>-0.351065</td>\n",
       "      <td>0.150541</td>\n",
       "      <td>-0.238958</td>\n",
       "      <td>-1.075925</td>\n",
       "      <td>-0.066461</td>\n",
       "      <td>0.145102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AACS</td>\n",
       "      <td>0.018061</td>\n",
       "      <td>0.611623</td>\n",
       "      <td>-0.300200</td>\n",
       "      <td>-0.821343</td>\n",
       "      <td>-0.486892</td>\n",
       "      <td>-0.306668</td>\n",
       "      <td>-0.465894</td>\n",
       "      <td>-0.297026</td>\n",
       "      <td>-1.019541</td>\n",
       "      <td>...</td>\n",
       "      <td>1.072971</td>\n",
       "      <td>1.265018</td>\n",
       "      <td>1.176459</td>\n",
       "      <td>0.958009</td>\n",
       "      <td>-0.074185</td>\n",
       "      <td>-0.419782</td>\n",
       "      <td>-0.959281</td>\n",
       "      <td>-1.004142</td>\n",
       "      <td>-1.073246</td>\n",
       "      <td>-1.126886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 233 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "0        A1BG   0.856487  -1.773695  -0.678679  -0.986914   \n",
       "1    A1BG-AS1   0.257664  -1.373085  -0.619923  -0.636275   \n",
       "2        A1CF  -0.089614  -0.546903   0.282914  -0.528926   \n",
       "3         A2M   0.552415  -0.635485  -0.954995  -0.259745   \n",
       "4       A2ML1   0.758031   1.549857   1.262225   1.338780   \n",
       "5     A3GALT2   0.435751   1.125971   1.479726   0.010520   \n",
       "6      A4GALT   1.277084   0.710266   0.272435   0.452544   \n",
       "7       A4GNT  -0.330903  -0.477462   0.267584  -0.426862   \n",
       "8        AAAS  -1.257280  -0.888406  -1.124108  -1.118677   \n",
       "9        AACS   0.018061   0.611623  -0.300200  -0.821343   \n",
       "\n",
       "   Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "0                    0.826986            0.948039   \n",
       "1                    0.362799            0.353296   \n",
       "2                    0.507916            0.577696   \n",
       "3                   -1.687391           -1.756847   \n",
       "4                   -0.289888           -0.407026   \n",
       "5                    0.877256            0.617134   \n",
       "6                    0.127867           -0.423357   \n",
       "7                    0.223556            0.260993   \n",
       "8                    1.234268            1.421483   \n",
       "9                   -0.486892           -0.306668   \n",
       "\n",
       "   Crus II, lateral hemisphere  Crus II, paravermis  Edinger-Westphal nucleus  \\\n",
       "0                     0.935427             1.120774                 -1.018554   \n",
       "1                     0.422766             0.346853                 -0.812015   \n",
       "2                     0.647671             0.306824                  0.089958   \n",
       "3                    -1.640242            -1.733110                 -0.091695   \n",
       "4                    -0.358798            -0.589988                  0.944684   \n",
       "5                     0.797456             0.724969                  0.692655   \n",
       "6                     0.222945            -0.453995                 -0.919359   \n",
       "7                     0.632328             0.097173                 -0.173487   \n",
       "8                     1.429758             1.603573                 -0.827011   \n",
       "9                    -0.465894            -0.297026                 -1.019541   \n",
       "\n",
       "   ...  temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "0  ...                        0.277830                      0.514923   \n",
       "1  ...                        1.074116                      0.821031   \n",
       "2  ...                       -0.030265                     -0.187367   \n",
       "3  ...                       -0.058505                      0.207109   \n",
       "4  ...                       -0.472908                     -0.598317   \n",
       "5  ...                       -0.339939                     -0.628313   \n",
       "6  ...                        0.100831                      0.322003   \n",
       "7  ...                        0.555680                     -0.462533   \n",
       "8  ...                        0.073441                     -0.372438   \n",
       "9  ...                        1.072971                      1.265018   \n",
       "\n",
       "   temporal pole, superior aspect  transverse gyri  trochlear nucleus  \\\n",
       "0                        0.733368        -0.104286          -0.910245   \n",
       "1                        1.219272         0.901213          -1.522431   \n",
       "2                       -0.428358        -0.465863          -0.136936   \n",
       "3                       -0.161808         0.183630           0.948098   \n",
       "4                       -0.247797        -0.282673           1.396365   \n",
       "5                       -0.246285         0.049359           0.223667   \n",
       "6                        0.622643         0.698657           0.060055   \n",
       "7                        0.752745         0.648017           1.639433   \n",
       "8                       -0.564808        -0.011762          -0.351065   \n",
       "9                        1.176459         0.958009          -0.074185   \n",
       "\n",
       "   tuberomammillary nucleus  ventral tegmental area  \\\n",
       "0                  1.039610               -0.155167   \n",
       "1                  0.598719               -1.709745   \n",
       "2                  1.229487               -0.110680   \n",
       "3                 -0.977692                0.911896   \n",
       "4                  0.945043                0.158202   \n",
       "5                  1.164986               -0.102258   \n",
       "6                 -1.546518               -0.578289   \n",
       "7                  0.353052                0.249200   \n",
       "8                  0.150541               -0.238958   \n",
       "9                 -0.419782               -0.959281   \n",
       "\n",
       "   ventromedial hypothalamic nucleus  vestibular nuclei  zona incerta  \n",
       "0                          -0.444398          -0.901361     -0.236790  \n",
       "1                          -0.054156          -1.695843     -1.155961  \n",
       "2                          -0.118175          -0.139776      0.123829  \n",
       "3                          -0.499357           1.469386      0.557998  \n",
       "4                           0.572771           0.073088     -0.886780  \n",
       "5                          -0.849872          -0.354230     -0.421125  \n",
       "6                          -1.117115          -0.442883     -0.352982  \n",
       "7                           0.934437          -0.165556     -0.715761  \n",
       "8                          -1.075925          -0.066461      0.145102  \n",
       "9                          -1.004142          -1.073246     -1.126886  \n",
       "\n",
       "[10 rows x 233 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the list of lists as a data frame\n",
    "# Note \"delimiter\" from loadtxt is \"sep\" here\n",
    "gene_df = pd.read_csv('Data/brainarea_vs_genes_exp_w_reannotations.tsv',sep='\\t')\n",
    "\n",
    "#Show the first five rows\n",
    "gene_df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the moment, the first column  of information above, called the **index**, just contains a list of numbers.\n",
    "\n",
    "> **Task**: Reassign the row labels by using the method `set_index()` on our dataframe. We can choose any column in our present dataframe to be the row values -- let's assign the row lables to be the `gene_symbol` and re-assign the dataframe. \n",
    "\n",
    "**Question for consideration**: Does `set_index()` work in place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "      <td>0.826986</td>\n",
       "      <td>0.948039</td>\n",
       "      <td>0.935427</td>\n",
       "      <td>1.120774</td>\n",
       "      <td>-1.018554</td>\n",
       "      <td>0.170282</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>-0.104286</td>\n",
       "      <td>-0.910245</td>\n",
       "      <td>1.039610</td>\n",
       "      <td>-0.155167</td>\n",
       "      <td>-0.444398</td>\n",
       "      <td>-0.901361</td>\n",
       "      <td>-0.236790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "      <td>0.362799</td>\n",
       "      <td>0.353296</td>\n",
       "      <td>0.422766</td>\n",
       "      <td>0.346853</td>\n",
       "      <td>-0.812015</td>\n",
       "      <td>0.903358</td>\n",
       "      <td>...</td>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.901213</td>\n",
       "      <td>-1.522431</td>\n",
       "      <td>0.598719</td>\n",
       "      <td>-1.709745</td>\n",
       "      <td>-0.054156</td>\n",
       "      <td>-1.695843</td>\n",
       "      <td>-1.155961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "      <td>0.507916</td>\n",
       "      <td>0.577696</td>\n",
       "      <td>0.647671</td>\n",
       "      <td>0.306824</td>\n",
       "      <td>0.089958</td>\n",
       "      <td>0.149820</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.465863</td>\n",
       "      <td>-0.136936</td>\n",
       "      <td>1.229487</td>\n",
       "      <td>-0.110680</td>\n",
       "      <td>-0.118175</td>\n",
       "      <td>-0.139776</td>\n",
       "      <td>0.123829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "      <td>-1.687391</td>\n",
       "      <td>-1.756847</td>\n",
       "      <td>-1.640242</td>\n",
       "      <td>-1.733110</td>\n",
       "      <td>-0.091695</td>\n",
       "      <td>0.003428</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.183630</td>\n",
       "      <td>0.948098</td>\n",
       "      <td>-0.977692</td>\n",
       "      <td>0.911896</td>\n",
       "      <td>-0.499357</td>\n",
       "      <td>1.469386</td>\n",
       "      <td>0.557998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "      <td>-0.289888</td>\n",
       "      <td>-0.407026</td>\n",
       "      <td>-0.358798</td>\n",
       "      <td>-0.589988</td>\n",
       "      <td>0.944684</td>\n",
       "      <td>-0.466327</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>-0.282673</td>\n",
       "      <td>1.396365</td>\n",
       "      <td>0.945043</td>\n",
       "      <td>0.158202</td>\n",
       "      <td>0.572771</td>\n",
       "      <td>0.073088</td>\n",
       "      <td>-0.886780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "gene_symbol                                               \n",
       "A1BG          0.856487  -1.773695  -0.678679  -0.986914   \n",
       "A1BG-AS1      0.257664  -1.373085  -0.619923  -0.636275   \n",
       "A1CF         -0.089614  -0.546903   0.282914  -0.528926   \n",
       "A2M           0.552415  -0.635485  -0.954995  -0.259745   \n",
       "A2ML1         0.758031   1.549857   1.262225   1.338780   \n",
       "\n",
       "             Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "gene_symbol                                                   \n",
       "A1BG                           0.826986            0.948039   \n",
       "A1BG-AS1                       0.362799            0.353296   \n",
       "A1CF                           0.507916            0.577696   \n",
       "A2M                           -1.687391           -1.756847   \n",
       "A2ML1                         -0.289888           -0.407026   \n",
       "\n",
       "             Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "gene_symbol                                                     \n",
       "A1BG                            0.935427             1.120774   \n",
       "A1BG-AS1                        0.422766             0.346853   \n",
       "A1CF                            0.647671             0.306824   \n",
       "A2M                            -1.640242            -1.733110   \n",
       "A2ML1                          -0.358798            -0.589988   \n",
       "\n",
       "             Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "gene_symbol                                            ...   \n",
       "A1BG                        -1.018554        0.170282  ...   \n",
       "A1BG-AS1                    -0.812015        0.903358  ...   \n",
       "A1CF                         0.089958        0.149820  ...   \n",
       "A2M                         -0.091695        0.003428  ...   \n",
       "A2ML1                        0.944684       -0.466327  ...   \n",
       "\n",
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  transverse gyri  \\\n",
       "gene_symbol                                                    \n",
       "A1BG                               0.733368        -0.104286   \n",
       "A1BG-AS1                           1.219272         0.901213   \n",
       "A1CF                              -0.428358        -0.465863   \n",
       "A2M                               -0.161808         0.183630   \n",
       "A2ML1                             -0.247797        -0.282673   \n",
       "\n",
       "             trochlear nucleus  tuberomammillary nucleus  \\\n",
       "gene_symbol                                                \n",
       "A1BG                 -0.910245                  1.039610   \n",
       "A1BG-AS1             -1.522431                  0.598719   \n",
       "A1CF                 -0.136936                  1.229487   \n",
       "A2M                   0.948098                 -0.977692   \n",
       "A2ML1                 1.396365                  0.945043   \n",
       "\n",
       "             ventral tegmental area  ventromedial hypothalamic nucleus  \\\n",
       "gene_symbol                                                              \n",
       "A1BG                      -0.155167                          -0.444398   \n",
       "A1BG-AS1                  -1.709745                          -0.054156   \n",
       "A1CF                      -0.110680                          -0.118175   \n",
       "A2M                        0.911896                          -0.499357   \n",
       "A2ML1                      0.158202                           0.572771   \n",
       "\n",
       "             vestibular nuclei  zona incerta  \n",
       "gene_symbol                                   \n",
       "A1BG                 -0.901361     -0.236790  \n",
       "A1BG-AS1             -1.695843     -1.155961  \n",
       "A1CF                 -0.139776      0.123829  \n",
       "A2M                   1.469386      0.557998  \n",
       "A2ML1                 0.073088     -0.886780  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index = 'gene_symbol'\n",
    "gene_df = gene_df.set_index(row_index)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would help to know what information is in our dataset. In other words, what is across the columns at the top?\n",
    "\n",
    "> **Task** Show a list of columns by accessing the `columns` attribute of our dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CA1 field', 'CA2 field', 'CA3 field', 'CA4 field',\n",
       "       'Crus I, lateral hemisphere', 'Crus I, paravermis',\n",
       "       'Crus II, lateral hemisphere', 'Crus II, paravermis',\n",
       "       'Edinger-Westphal nucleus', 'Heschl's gyrus',\n",
       "       ...\n",
       "       'temporal pole, inferior aspect', 'temporal pole, medial aspect',\n",
       "       'temporal pole, superior aspect', 'transverse gyri',\n",
       "       'trochlear nucleus', 'tuberomammillary nucleus',\n",
       "       'ventral tegmental area', 'ventromedial hypothalamic nucleus',\n",
       "       'vestibular nuclei', 'zona incerta'],\n",
       "      dtype='object', length=232)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Access the columns of our dataframe \n",
    "columns = gene_df.columns\n",
    "columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Indexing in Pandas works slightly different than in NumPy. Similar to a dictionary, we can index dataframes by their names. \n",
    "\n",
    "The syntax for indexing **single locations** in a dataframe is `dataframe.loc[row_label,column_label]`.\n",
    "\n",
    "To index an **individual column**, we use the shorthand syntax `dataframe[column_label]`.\n",
    "\n",
    "To index an **individual row**, we use the syntax `dataframe.loc[row_label]`.\n",
    "\n",
    "To index by **index #**, we use the syntax `dataframe.iloc[index_number]`.\n",
    "\n",
    "Below are some examples on how to access rows, columns, and single values in our dataframe. For more information on indexing dataframes, visit the <a href = \"https://pandas.pydata.org/docs/user_guide/indexing.html#indexing\"> \"Indexing and selecting data\"</a> section in the Pandas User Guide.\n",
    "\n",
    "> **Task**: Try each of these forms of indexing below. Can you use this syntax to get multiple values (a slice or subset) of the dataframe as well?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8564873784944677"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single location\n",
    "gene_df.loc['A1BG','CA1 field']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_symbol\n",
       "A1BG       -0.155167\n",
       "A1BG-AS1   -1.709745\n",
       "A1CF       -0.110680\n",
       "A2M         0.911896\n",
       "A2ML1       0.158202\n",
       "              ...   \n",
       "ZYG11A     -0.229960\n",
       "ZYG11B     -0.828416\n",
       "ZYX        -0.411662\n",
       "ZZEF1       0.807997\n",
       "ZZZ3        0.694799\n",
       "Name: ventral tegmental area, Length: 20869, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single column\n",
    "gene_df['ventral tegmental area']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.522811\n",
       "CA2 field                            0.425057\n",
       "CA3 field                            0.317641\n",
       "CA4 field                            1.039662\n",
       "Crus I, lateral hemisphere          -1.272901\n",
       "                                       ...   \n",
       "tuberomammillary nucleus            -1.086945\n",
       "ventral tegmental area              -0.702538\n",
       "ventromedial hypothalamic nucleus   -0.485458\n",
       "vestibular nuclei                   -1.078950\n",
       "zona incerta                         0.854158\n",
       "Name: ARX, Length: 232, dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single row\n",
    "gene_df.loc['ARX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9079172385353984"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select a single index\n",
    "# Hint: You can show all indices using gene_df.index\n",
    "gene_df.iloc[10001,0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To select multiple different columns, you can use a `list` of all your columns of interest as so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field\n",
       "gene_symbol                                            \n",
       "A1BG          0.856487  -1.773695  -0.678679  -0.986914\n",
       "A1BG-AS1      0.257664  -1.373085  -0.619923  -0.636275\n",
       "A1CF         -0.089614  -0.546903   0.282914  -0.528926\n",
       "A2M           0.552415  -0.635485  -0.954995  -0.259745\n",
       "A2ML1         0.758031   1.549857   1.262225   1.338780"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe w/ only CA field entries \n",
    "CA_field_df = gene_df[['CA1 field', \n",
    "                       'CA2 field', \n",
    "                       'CA3 field', \n",
    "                       'CA4 field']]\n",
    "CA_field_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsetting dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like NumPy arrays, we can subset our original dataframe to only include data that meets our criteria. Our dataframe has data on multiple different brain areas with many gene expression values. You can filter this dataframe using the following syntax:\n",
    "```\n",
    "new_df = original_df[original_df['Column of Interest'] == 'Desired Value']\n",
    "```\n",
    "In plain english, what this is saying is: save a dataframe from the original dataframe, where the original dataframe values in my Column of Interest are equal to my Desired Value. For more information on subsetting,  visit the <a href = \"https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html\"> \"How do I select a subset of a DataFrame\"</a> section in the Pandas documentation. \n",
    "\n",
    "Below we will demonstrate how to execute this by taking a look at the `CA1 field` column in `gene_df`. We will create a dataframe from `gene_df` that only contains genes that showed a certain level of gene expression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCC12</th>\n",
       "      <td>2.089999</td>\n",
       "      <td>0.684837</td>\n",
       "      <td>0.097313</td>\n",
       "      <td>-0.051411</td>\n",
       "      <td>-1.078900</td>\n",
       "      <td>-0.912071</td>\n",
       "      <td>-1.131497</td>\n",
       "      <td>-0.799075</td>\n",
       "      <td>-0.009423</td>\n",
       "      <td>0.889132</td>\n",
       "      <td>...</td>\n",
       "      <td>1.289340</td>\n",
       "      <td>0.885886</td>\n",
       "      <td>1.271053</td>\n",
       "      <td>0.650801</td>\n",
       "      <td>-0.083413</td>\n",
       "      <td>-0.793237</td>\n",
       "      <td>-0.499512</td>\n",
       "      <td>-0.762330</td>\n",
       "      <td>-0.902496</td>\n",
       "      <td>-0.904421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABHD17C</th>\n",
       "      <td>1.716973</td>\n",
       "      <td>0.601041</td>\n",
       "      <td>1.132500</td>\n",
       "      <td>1.354679</td>\n",
       "      <td>-0.923195</td>\n",
       "      <td>-0.887576</td>\n",
       "      <td>-1.122027</td>\n",
       "      <td>-0.926876</td>\n",
       "      <td>0.339678</td>\n",
       "      <td>0.208174</td>\n",
       "      <td>...</td>\n",
       "      <td>0.530983</td>\n",
       "      <td>0.809101</td>\n",
       "      <td>0.763170</td>\n",
       "      <td>-0.093598</td>\n",
       "      <td>-0.313468</td>\n",
       "      <td>-0.164013</td>\n",
       "      <td>-0.537971</td>\n",
       "      <td>1.169105</td>\n",
       "      <td>-0.663245</td>\n",
       "      <td>-0.377181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>2.051762</td>\n",
       "      <td>2.571777</td>\n",
       "      <td>2.472188</td>\n",
       "      <td>2.261170</td>\n",
       "      <td>0.366138</td>\n",
       "      <td>0.507783</td>\n",
       "      <td>0.449606</td>\n",
       "      <td>0.498424</td>\n",
       "      <td>-1.365242</td>\n",
       "      <td>0.472104</td>\n",
       "      <td>...</td>\n",
       "      <td>0.868471</td>\n",
       "      <td>1.197022</td>\n",
       "      <td>0.914783</td>\n",
       "      <td>0.473846</td>\n",
       "      <td>-2.230209</td>\n",
       "      <td>-0.330684</td>\n",
       "      <td>-1.153189</td>\n",
       "      <td>-0.073650</td>\n",
       "      <td>-1.681602</td>\n",
       "      <td>-1.049258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTB</th>\n",
       "      <td>2.489711</td>\n",
       "      <td>2.806688</td>\n",
       "      <td>2.461655</td>\n",
       "      <td>2.340131</td>\n",
       "      <td>-1.296731</td>\n",
       "      <td>-1.334696</td>\n",
       "      <td>-1.158460</td>\n",
       "      <td>-1.461027</td>\n",
       "      <td>0.088209</td>\n",
       "      <td>0.027236</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.370215</td>\n",
       "      <td>-0.946920</td>\n",
       "      <td>-0.197363</td>\n",
       "      <td>-0.094468</td>\n",
       "      <td>0.926183</td>\n",
       "      <td>-0.034337</td>\n",
       "      <td>0.803164</td>\n",
       "      <td>-0.389238</td>\n",
       "      <td>0.596916</td>\n",
       "      <td>0.110921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTR2</th>\n",
       "      <td>2.655049</td>\n",
       "      <td>2.384445</td>\n",
       "      <td>1.728348</td>\n",
       "      <td>1.413585</td>\n",
       "      <td>-1.377474</td>\n",
       "      <td>-1.207922</td>\n",
       "      <td>-1.496018</td>\n",
       "      <td>-1.359462</td>\n",
       "      <td>0.632764</td>\n",
       "      <td>0.259402</td>\n",
       "      <td>...</td>\n",
       "      <td>0.215683</td>\n",
       "      <td>0.142880</td>\n",
       "      <td>0.080649</td>\n",
       "      <td>0.312218</td>\n",
       "      <td>1.793563</td>\n",
       "      <td>-0.028263</td>\n",
       "      <td>-0.189930</td>\n",
       "      <td>-0.389064</td>\n",
       "      <td>-0.484580</td>\n",
       "      <td>-1.478232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF878</th>\n",
       "      <td>1.726832</td>\n",
       "      <td>1.428443</td>\n",
       "      <td>0.755283</td>\n",
       "      <td>0.095314</td>\n",
       "      <td>1.302215</td>\n",
       "      <td>1.357688</td>\n",
       "      <td>1.359651</td>\n",
       "      <td>1.186241</td>\n",
       "      <td>-1.360845</td>\n",
       "      <td>-0.354530</td>\n",
       "      <td>...</td>\n",
       "      <td>0.538528</td>\n",
       "      <td>0.226543</td>\n",
       "      <td>-0.110654</td>\n",
       "      <td>-1.117464</td>\n",
       "      <td>1.480377</td>\n",
       "      <td>-0.128002</td>\n",
       "      <td>-0.229289</td>\n",
       "      <td>-2.460564</td>\n",
       "      <td>-0.641011</td>\n",
       "      <td>-1.396994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF883</th>\n",
       "      <td>2.113267</td>\n",
       "      <td>2.247842</td>\n",
       "      <td>2.314902</td>\n",
       "      <td>1.906474</td>\n",
       "      <td>-0.707426</td>\n",
       "      <td>-0.706865</td>\n",
       "      <td>-0.658908</td>\n",
       "      <td>-0.782010</td>\n",
       "      <td>-3.108014</td>\n",
       "      <td>0.146046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.666075</td>\n",
       "      <td>1.114869</td>\n",
       "      <td>0.936464</td>\n",
       "      <td>0.194066</td>\n",
       "      <td>-0.757217</td>\n",
       "      <td>-0.543755</td>\n",
       "      <td>-0.370543</td>\n",
       "      <td>0.363171</td>\n",
       "      <td>-1.222364</td>\n",
       "      <td>0.120239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF4</th>\n",
       "      <td>2.392668</td>\n",
       "      <td>3.334372</td>\n",
       "      <td>2.680485</td>\n",
       "      <td>2.376334</td>\n",
       "      <td>0.109822</td>\n",
       "      <td>-0.064917</td>\n",
       "      <td>-0.169874</td>\n",
       "      <td>-0.069484</td>\n",
       "      <td>0.830355</td>\n",
       "      <td>-0.467305</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.218960</td>\n",
       "      <td>0.432994</td>\n",
       "      <td>0.387828</td>\n",
       "      <td>-0.627475</td>\n",
       "      <td>1.119143</td>\n",
       "      <td>-1.085945</td>\n",
       "      <td>-0.141248</td>\n",
       "      <td>-0.222854</td>\n",
       "      <td>0.169401</td>\n",
       "      <td>0.101495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN5A</th>\n",
       "      <td>1.828036</td>\n",
       "      <td>1.469749</td>\n",
       "      <td>0.682074</td>\n",
       "      <td>-0.511445</td>\n",
       "      <td>-0.606073</td>\n",
       "      <td>-0.638284</td>\n",
       "      <td>-0.523978</td>\n",
       "      <td>-0.805322</td>\n",
       "      <td>1.128197</td>\n",
       "      <td>-0.430763</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.085916</td>\n",
       "      <td>0.628877</td>\n",
       "      <td>-0.136047</td>\n",
       "      <td>-0.497448</td>\n",
       "      <td>3.023357</td>\n",
       "      <td>1.014005</td>\n",
       "      <td>0.199544</td>\n",
       "      <td>-0.475723</td>\n",
       "      <td>0.339935</td>\n",
       "      <td>-0.371764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZWINT</th>\n",
       "      <td>3.154188</td>\n",
       "      <td>3.326311</td>\n",
       "      <td>2.876940</td>\n",
       "      <td>1.813612</td>\n",
       "      <td>-1.054751</td>\n",
       "      <td>-1.103545</td>\n",
       "      <td>-1.080372</td>\n",
       "      <td>-1.171860</td>\n",
       "      <td>-1.682247</td>\n",
       "      <td>0.463577</td>\n",
       "      <td>...</td>\n",
       "      <td>0.605577</td>\n",
       "      <td>0.920985</td>\n",
       "      <td>1.071082</td>\n",
       "      <td>0.519482</td>\n",
       "      <td>1.304839</td>\n",
       "      <td>1.645788</td>\n",
       "      <td>-0.402724</td>\n",
       "      <td>-1.210645</td>\n",
       "      <td>-0.818001</td>\n",
       "      <td>-0.473278</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "gene_symbol                                               \n",
       "ABCC12        2.089999   0.684837   0.097313  -0.051411   \n",
       "ABHD17C       1.716973   0.601041   1.132500   1.354679   \n",
       "ABI1          2.051762   2.571777   2.472188   2.261170   \n",
       "ACTB          2.489711   2.806688   2.461655   2.340131   \n",
       "ACTR2         2.655049   2.384445   1.728348   1.413585   \n",
       "...                ...        ...        ...        ...   \n",
       "ZNF878        1.726832   1.428443   0.755283   0.095314   \n",
       "ZNF883        2.113267   2.247842   2.314902   1.906474   \n",
       "ZNRF4         2.392668   3.334372   2.680485   2.376334   \n",
       "ZSCAN5A       1.828036   1.469749   0.682074  -0.511445   \n",
       "ZWINT         3.154188   3.326311   2.876940   1.813612   \n",
       "\n",
       "             Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "gene_symbol                                                   \n",
       "ABCC12                        -1.078900           -0.912071   \n",
       "ABHD17C                       -0.923195           -0.887576   \n",
       "ABI1                           0.366138            0.507783   \n",
       "ACTB                          -1.296731           -1.334696   \n",
       "ACTR2                         -1.377474           -1.207922   \n",
       "...                                 ...                 ...   \n",
       "ZNF878                         1.302215            1.357688   \n",
       "ZNF883                        -0.707426           -0.706865   \n",
       "ZNRF4                          0.109822           -0.064917   \n",
       "ZSCAN5A                       -0.606073           -0.638284   \n",
       "ZWINT                         -1.054751           -1.103545   \n",
       "\n",
       "             Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "gene_symbol                                                     \n",
       "ABCC12                         -1.131497            -0.799075   \n",
       "ABHD17C                        -1.122027            -0.926876   \n",
       "ABI1                            0.449606             0.498424   \n",
       "ACTB                           -1.158460            -1.461027   \n",
       "ACTR2                          -1.496018            -1.359462   \n",
       "...                                  ...                  ...   \n",
       "ZNF878                          1.359651             1.186241   \n",
       "ZNF883                         -0.658908            -0.782010   \n",
       "ZNRF4                          -0.169874            -0.069484   \n",
       "ZSCAN5A                        -0.523978            -0.805322   \n",
       "ZWINT                          -1.080372            -1.171860   \n",
       "\n",
       "             Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "gene_symbol                                            ...   \n",
       "ABCC12                      -0.009423        0.889132  ...   \n",
       "ABHD17C                      0.339678        0.208174  ...   \n",
       "ABI1                        -1.365242        0.472104  ...   \n",
       "ACTB                         0.088209        0.027236  ...   \n",
       "ACTR2                        0.632764        0.259402  ...   \n",
       "...                               ...             ...  ...   \n",
       "ZNF878                      -1.360845       -0.354530  ...   \n",
       "ZNF883                      -3.108014        0.146046  ...   \n",
       "ZNRF4                        0.830355       -0.467305  ...   \n",
       "ZSCAN5A                      1.128197       -0.430763  ...   \n",
       "ZWINT                       -1.682247        0.463577  ...   \n",
       "\n",
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "ABCC12                             1.289340                      0.885886   \n",
       "ABHD17C                            0.530983                      0.809101   \n",
       "ABI1                               0.868471                      1.197022   \n",
       "ACTB                              -0.370215                     -0.946920   \n",
       "ACTR2                              0.215683                      0.142880   \n",
       "...                                     ...                           ...   \n",
       "ZNF878                             0.538528                      0.226543   \n",
       "ZNF883                             0.666075                      1.114869   \n",
       "ZNRF4                             -0.218960                      0.432994   \n",
       "ZSCAN5A                           -0.085916                      0.628877   \n",
       "ZWINT                              0.605577                      0.920985   \n",
       "\n",
       "             temporal pole, superior aspect  transverse gyri  \\\n",
       "gene_symbol                                                    \n",
       "ABCC12                             1.271053         0.650801   \n",
       "ABHD17C                            0.763170        -0.093598   \n",
       "ABI1                               0.914783         0.473846   \n",
       "ACTB                              -0.197363        -0.094468   \n",
       "ACTR2                              0.080649         0.312218   \n",
       "...                                     ...              ...   \n",
       "ZNF878                            -0.110654        -1.117464   \n",
       "ZNF883                             0.936464         0.194066   \n",
       "ZNRF4                              0.387828        -0.627475   \n",
       "ZSCAN5A                           -0.136047        -0.497448   \n",
       "ZWINT                              1.071082         0.519482   \n",
       "\n",
       "             trochlear nucleus  tuberomammillary nucleus  \\\n",
       "gene_symbol                                                \n",
       "ABCC12               -0.083413                 -0.793237   \n",
       "ABHD17C              -0.313468                 -0.164013   \n",
       "ABI1                 -2.230209                 -0.330684   \n",
       "ACTB                  0.926183                 -0.034337   \n",
       "ACTR2                 1.793563                 -0.028263   \n",
       "...                        ...                       ...   \n",
       "ZNF878                1.480377                 -0.128002   \n",
       "ZNF883               -0.757217                 -0.543755   \n",
       "ZNRF4                 1.119143                 -1.085945   \n",
       "ZSCAN5A               3.023357                  1.014005   \n",
       "ZWINT                 1.304839                  1.645788   \n",
       "\n",
       "             ventral tegmental area  ventromedial hypothalamic nucleus  \\\n",
       "gene_symbol                                                              \n",
       "ABCC12                    -0.499512                          -0.762330   \n",
       "ABHD17C                   -0.537971                           1.169105   \n",
       "ABI1                      -1.153189                          -0.073650   \n",
       "ACTB                       0.803164                          -0.389238   \n",
       "ACTR2                     -0.189930                          -0.389064   \n",
       "...                             ...                                ...   \n",
       "ZNF878                    -0.229289                          -2.460564   \n",
       "ZNF883                    -0.370543                           0.363171   \n",
       "ZNRF4                     -0.141248                          -0.222854   \n",
       "ZSCAN5A                    0.199544                          -0.475723   \n",
       "ZWINT                     -0.402724                          -1.210645   \n",
       "\n",
       "             vestibular nuclei  zona incerta  \n",
       "gene_symbol                                   \n",
       "ABCC12               -0.902496     -0.904421  \n",
       "ABHD17C              -0.663245     -0.377181  \n",
       "ABI1                 -1.681602     -1.049258  \n",
       "ACTB                  0.596916      0.110921  \n",
       "ACTR2                -0.484580     -1.478232  \n",
       "...                        ...           ...  \n",
       "ZNF878               -0.641011     -1.396994  \n",
       "ZNF883               -1.222364      0.120239  \n",
       "ZNRF4                 0.169401      0.101495  \n",
       "ZSCAN5A               0.339935     -0.371764  \n",
       "ZWINT                -0.818001     -0.473278  \n",
       "\n",
       "[772 rows x 232 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a dataframe with only genes that have an expression \n",
    "# value greater than 1.7 in 'CA1 field' \n",
    "\n",
    "high_CA1 = gene_df[gene_df['CA1 field'] > 1.7]\n",
    "high_CA1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DataFrame Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas has many useful methods that you can use on your data, including `describe`, `mean`, and more. To learn more about all the different methods that can be used to manipulate and analyze dataframes, please visit the <a href = \"https://pandas.pydata.org/docs/user_guide/index.html\"> Pandas User Guide </a>. We will demonstrate some of these methods below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `describe` method returns descriptive statistics of all the columns in our dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "      <td>20869.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.003664</td>\n",
       "      <td>0.017002</td>\n",
       "      <td>0.015315</td>\n",
       "      <td>-0.016633</td>\n",
       "      <td>0.093686</td>\n",
       "      <td>0.088810</td>\n",
       "      <td>0.096118</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>0.047124</td>\n",
       "      <td>-0.042517</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.056050</td>\n",
       "      <td>-0.051731</td>\n",
       "      <td>-0.049905</td>\n",
       "      <td>-0.039528</td>\n",
       "      <td>0.059726</td>\n",
       "      <td>0.014856</td>\n",
       "      <td>0.009535</td>\n",
       "      <td>-0.002853</td>\n",
       "      <td>-0.002527</td>\n",
       "      <td>0.013018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.924456</td>\n",
       "      <td>1.129368</td>\n",
       "      <td>1.078987</td>\n",
       "      <td>0.897192</td>\n",
       "      <td>1.146146</td>\n",
       "      <td>1.118501</td>\n",
       "      <td>1.172986</td>\n",
       "      <td>1.136823</td>\n",
       "      <td>0.973224</td>\n",
       "      <td>0.500526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.567950</td>\n",
       "      <td>0.651098</td>\n",
       "      <td>0.636729</td>\n",
       "      <td>0.494012</td>\n",
       "      <td>1.158916</td>\n",
       "      <td>0.897387</td>\n",
       "      <td>0.686320</td>\n",
       "      <td>0.830602</td>\n",
       "      <td>0.723977</td>\n",
       "      <td>0.725770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-4.076424</td>\n",
       "      <td>-5.923691</td>\n",
       "      <td>-5.994731</td>\n",
       "      <td>-3.971984</td>\n",
       "      <td>-2.739924</td>\n",
       "      <td>-2.662897</td>\n",
       "      <td>-2.908676</td>\n",
       "      <td>-2.864308</td>\n",
       "      <td>-3.671242</td>\n",
       "      <td>-1.666268</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.840486</td>\n",
       "      <td>-2.433961</td>\n",
       "      <td>-2.412614</td>\n",
       "      <td>-1.655962</td>\n",
       "      <td>-6.330275</td>\n",
       "      <td>-3.141490</td>\n",
       "      <td>-1.977225</td>\n",
       "      <td>-3.541112</td>\n",
       "      <td>-2.369304</td>\n",
       "      <td>-2.348784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-0.570475</td>\n",
       "      <td>-0.644093</td>\n",
       "      <td>-0.631248</td>\n",
       "      <td>-0.573605</td>\n",
       "      <td>-0.802651</td>\n",
       "      <td>-0.778933</td>\n",
       "      <td>-0.824404</td>\n",
       "      <td>-0.794683</td>\n",
       "      <td>-0.651691</td>\n",
       "      <td>-0.414882</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.472481</td>\n",
       "      <td>-0.517562</td>\n",
       "      <td>-0.529458</td>\n",
       "      <td>-0.404416</td>\n",
       "      <td>-0.751329</td>\n",
       "      <td>-0.602619</td>\n",
       "      <td>-0.497988</td>\n",
       "      <td>-0.542164</td>\n",
       "      <td>-0.532177</td>\n",
       "      <td>-0.493921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>-0.025821</td>\n",
       "      <td>0.011189</td>\n",
       "      <td>0.006140</td>\n",
       "      <td>-0.044566</td>\n",
       "      <td>0.100558</td>\n",
       "      <td>0.098665</td>\n",
       "      <td>0.109358</td>\n",
       "      <td>0.096585</td>\n",
       "      <td>0.022760</td>\n",
       "      <td>-0.057495</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.091029</td>\n",
       "      <td>-0.083971</td>\n",
       "      <td>-0.089572</td>\n",
       "      <td>-0.051758</td>\n",
       "      <td>0.037410</td>\n",
       "      <td>-0.030474</td>\n",
       "      <td>-0.023799</td>\n",
       "      <td>-0.023471</td>\n",
       "      <td>-0.045003</td>\n",
       "      <td>-0.031623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.561571</td>\n",
       "      <td>0.706946</td>\n",
       "      <td>0.706398</td>\n",
       "      <td>0.547903</td>\n",
       "      <td>0.985159</td>\n",
       "      <td>0.951919</td>\n",
       "      <td>1.007791</td>\n",
       "      <td>0.960813</td>\n",
       "      <td>0.736614</td>\n",
       "      <td>0.312739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.352796</td>\n",
       "      <td>0.412321</td>\n",
       "      <td>0.403382</td>\n",
       "      <td>0.310551</td>\n",
       "      <td>0.844723</td>\n",
       "      <td>0.587382</td>\n",
       "      <td>0.503613</td>\n",
       "      <td>0.508410</td>\n",
       "      <td>0.501191</td>\n",
       "      <td>0.492262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>7.062717</td>\n",
       "      <td>7.387742</td>\n",
       "      <td>6.413603</td>\n",
       "      <td>7.178692</td>\n",
       "      <td>2.679149</td>\n",
       "      <td>2.717237</td>\n",
       "      <td>2.963899</td>\n",
       "      <td>2.857205</td>\n",
       "      <td>7.552440</td>\n",
       "      <td>2.234669</td>\n",
       "      <td>...</td>\n",
       "      <td>2.199291</td>\n",
       "      <td>2.631498</td>\n",
       "      <td>3.065735</td>\n",
       "      <td>2.238555</td>\n",
       "      <td>6.892682</td>\n",
       "      <td>5.968364</td>\n",
       "      <td>7.267837</td>\n",
       "      <td>6.650673</td>\n",
       "      <td>2.723777</td>\n",
       "      <td>2.845665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CA1 field     CA2 field     CA3 field     CA4 field  \\\n",
       "count  20869.000000  20869.000000  20869.000000  20869.000000   \n",
       "mean       0.003664      0.017002      0.015315     -0.016633   \n",
       "std        0.924456      1.129368      1.078987      0.897192   \n",
       "min       -4.076424     -5.923691     -5.994731     -3.971984   \n",
       "25%       -0.570475     -0.644093     -0.631248     -0.573605   \n",
       "50%       -0.025821      0.011189      0.006140     -0.044566   \n",
       "75%        0.561571      0.706946      0.706398      0.547903   \n",
       "max        7.062717      7.387742      6.413603      7.178692   \n",
       "\n",
       "       Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "count                20869.000000        20869.000000   \n",
       "mean                     0.093686            0.088810   \n",
       "std                      1.146146            1.118501   \n",
       "min                     -2.739924           -2.662897   \n",
       "25%                     -0.802651           -0.778933   \n",
       "50%                      0.100558            0.098665   \n",
       "75%                      0.985159            0.951919   \n",
       "max                      2.679149            2.717237   \n",
       "\n",
       "       Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "count                 20869.000000         20869.000000   \n",
       "mean                      0.096118             0.087608   \n",
       "std                       1.172986             1.136823   \n",
       "min                      -2.908676            -2.864308   \n",
       "25%                      -0.824404            -0.794683   \n",
       "50%                       0.109358             0.096585   \n",
       "75%                       1.007791             0.960813   \n",
       "max                       2.963899             2.857205   \n",
       "\n",
       "       Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "count              20869.000000    20869.000000  ...   \n",
       "mean                   0.047124       -0.042517  ...   \n",
       "std                    0.973224        0.500526  ...   \n",
       "min                   -3.671242       -1.666268  ...   \n",
       "25%                   -0.651691       -0.414882  ...   \n",
       "50%                    0.022760       -0.057495  ...   \n",
       "75%                    0.736614        0.312739  ...   \n",
       "max                    7.552440        2.234669  ...   \n",
       "\n",
       "       temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "count                    20869.000000                  20869.000000   \n",
       "mean                        -0.056050                     -0.051731   \n",
       "std                          0.567950                      0.651098   \n",
       "min                         -1.840486                     -2.433961   \n",
       "25%                         -0.472481                     -0.517562   \n",
       "50%                         -0.091029                     -0.083971   \n",
       "75%                          0.352796                      0.412321   \n",
       "max                          2.199291                      2.631498   \n",
       "\n",
       "       temporal pole, superior aspect  transverse gyri  trochlear nucleus  \\\n",
       "count                    20869.000000     20869.000000       20869.000000   \n",
       "mean                        -0.049905        -0.039528           0.059726   \n",
       "std                          0.636729         0.494012           1.158916   \n",
       "min                         -2.412614        -1.655962          -6.330275   \n",
       "25%                         -0.529458        -0.404416          -0.751329   \n",
       "50%                         -0.089572        -0.051758           0.037410   \n",
       "75%                          0.403382         0.310551           0.844723   \n",
       "max                          3.065735         2.238555           6.892682   \n",
       "\n",
       "       tuberomammillary nucleus  ventral tegmental area  \\\n",
       "count              20869.000000            20869.000000   \n",
       "mean                   0.014856                0.009535   \n",
       "std                    0.897387                0.686320   \n",
       "min                   -3.141490               -1.977225   \n",
       "25%                   -0.602619               -0.497988   \n",
       "50%                   -0.030474               -0.023799   \n",
       "75%                    0.587382                0.503613   \n",
       "max                    5.968364                7.267837   \n",
       "\n",
       "       ventromedial hypothalamic nucleus  vestibular nuclei  zona incerta  \n",
       "count                       20869.000000       20869.000000  20869.000000  \n",
       "mean                           -0.002853          -0.002527      0.013018  \n",
       "std                             0.830602           0.723977      0.725770  \n",
       "min                            -3.541112          -2.369304     -2.348784  \n",
       "25%                            -0.542164          -0.532177     -0.493921  \n",
       "50%                            -0.023471          -0.045003     -0.031623  \n",
       "75%                             0.508410           0.501191      0.492262  \n",
       "max                             6.650673           2.723777      2.845665  \n",
       "\n",
       "[8 rows x 232 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `mean` and `std` method return the mean and standard deviation of each column in the dataframe, respectfully. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.003664\n",
       "CA2 field                            0.017002\n",
       "CA3 field                            0.015315\n",
       "CA4 field                           -0.016633\n",
       "Crus I, lateral hemisphere           0.093686\n",
       "                                       ...   \n",
       "tuberomammillary nucleus             0.014856\n",
       "ventral tegmental area               0.009535\n",
       "ventromedial hypothalamic nucleus   -0.002853\n",
       "vestibular nuclei                   -0.002527\n",
       "zona incerta                         0.013018\n",
       "Length: 232, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try df.mean()\n",
    "gene_df.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CA1 field                            0.924456\n",
       "CA2 field                            1.129368\n",
       "CA3 field                            1.078987\n",
       "CA4 field                            0.897192\n",
       "Crus I, lateral hemisphere           1.146146\n",
       "                                       ...   \n",
       "tuberomammillary nucleus             0.897387\n",
       "ventral tegmental area               0.686320\n",
       "ventromedial hypothalamic nucleus    0.830602\n",
       "vestibular nuclei                    0.723977\n",
       "zona incerta                         0.725770\n",
       "Length: 232, dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Try df.std()\n",
    "gene_df.std()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging & joining dataframes\n",
    "\n",
    "Let's say we have two different dataframes and we would like to combine the two into one single dataframe. We can use either the `merge` or `join` Pandas methods in order to pull all of this data into one dataframe. \n",
    "\n",
    "![](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png)\n",
    "\n",
    "There are different types of joins/merges you can do in Pandas, illustrated <a href=\"http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/\">above</a>. Here, we want to do an **inner** merge, where we're only keeping entries with indices that are in both dataframes. We could do this merge based on columns, alternatively.\n",
    "\n",
    "**Inner** is the default kind of join, so we do not need to specify it. And by default, join will use the 'left' dataframe, in other words, the dataframe that is executing the `join` method.\n",
    "\n",
    "If you need more information, look at the <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html\">join</a> and <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html\">merge</a> documentation: you can use either of these to unite your dataframes, though join will be simpler!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is an example of how to join two separate dataframe into one, unified dataframe. We start with one dataframe with only entries from the *temporal pole* and another dataframe with only entries from the CA fields of the hippocampus. We can then join the two dataframes together using the syntax `unified_df = df_1.join(df_2)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  \n",
       "gene_symbol                                  \n",
       "A1BG                               0.733368  \n",
       "A1BG-AS1                           1.219272  \n",
       "A1CF                              -0.428358  \n",
       "A2M                               -0.161808  \n",
       "A2ML1                             -0.247797  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe w/ only Temporal Pole entries \n",
    "temporal_pole_df = gene_df[['temporal pole, inferior aspect', \n",
    "                            'temporal pole, medial aspect', \n",
    "                            'temporal pole, superior aspect']]\n",
    "temporal_pole_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.277830</td>\n",
       "      <td>0.514923</td>\n",
       "      <td>0.733368</td>\n",
       "      <td>0.856487</td>\n",
       "      <td>-1.773695</td>\n",
       "      <td>-0.678679</td>\n",
       "      <td>-0.986914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>1.074116</td>\n",
       "      <td>0.821031</td>\n",
       "      <td>1.219272</td>\n",
       "      <td>0.257664</td>\n",
       "      <td>-1.373085</td>\n",
       "      <td>-0.619923</td>\n",
       "      <td>-0.636275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-0.030265</td>\n",
       "      <td>-0.187367</td>\n",
       "      <td>-0.428358</td>\n",
       "      <td>-0.089614</td>\n",
       "      <td>-0.546903</td>\n",
       "      <td>0.282914</td>\n",
       "      <td>-0.528926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-0.058505</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>-0.161808</td>\n",
       "      <td>0.552415</td>\n",
       "      <td>-0.635485</td>\n",
       "      <td>-0.954995</td>\n",
       "      <td>-0.259745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>-0.472908</td>\n",
       "      <td>-0.598317</td>\n",
       "      <td>-0.247797</td>\n",
       "      <td>0.758031</td>\n",
       "      <td>1.549857</td>\n",
       "      <td>1.262225</td>\n",
       "      <td>1.338780</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "A1BG                               0.277830                      0.514923   \n",
       "A1BG-AS1                           1.074116                      0.821031   \n",
       "A1CF                              -0.030265                     -0.187367   \n",
       "A2M                               -0.058505                      0.207109   \n",
       "A2ML1                             -0.472908                     -0.598317   \n",
       "\n",
       "             temporal pole, superior aspect  CA1 field  CA2 field  CA3 field  \\\n",
       "gene_symbol                                                                    \n",
       "A1BG                               0.733368   0.856487  -1.773695  -0.678679   \n",
       "A1BG-AS1                           1.219272   0.257664  -1.373085  -0.619923   \n",
       "A1CF                              -0.428358  -0.089614  -0.546903   0.282914   \n",
       "A2M                               -0.161808   0.552415  -0.635485  -0.954995   \n",
       "A2ML1                             -0.247797   0.758031   1.549857   1.262225   \n",
       "\n",
       "             CA4 field  \n",
       "gene_symbol             \n",
       "A1BG         -0.986914  \n",
       "A1BG-AS1     -0.636275  \n",
       "A1CF         -0.528926  \n",
       "A2M          -0.259745  \n",
       "A2ML1         1.338780  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the two dataframes\n",
    "df_1 = temporal_pole_df\n",
    "df_2 = CA_field_df\n",
    "\n",
    "unified_df = df_1.join(df_2)\n",
    "unified_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "## Group Task (To submit on Canvas)\n",
    "### *Does the number of neurons in the cortex scale with cortex size?*\n",
    "In this final task, we'll look at [this dataset](http://www.suzanaherculanohouzel.com/2015-h-h-et-al-bbe-database/), which is already uploaded to your DataHub, to figure out whether the number of neurons in the cortex scales with cortex size, and if this scaling is different for primates.  \n",
    "\n",
    "To complete this task, do the following steps:\n",
    "\n",
    "1. Use `pd.read_csv()` to read in our comma-delimited file 'Data/species_brainmass_neurons.csv' and assign it to a dataframe called `brains_df`.\n",
    "2. Separate the dataframe into two dataframes: one that contains all of the primates (`primate_df`; 'Order' is Primata) and one that contains all of the data that are *not* from primates (`nonprimate_df`). **Hint**: You can subset the dataframe by using the syntax `brains_df[brains_df['Order']=='Primata']`.\n",
    "3. Using two separate, subsequent calls to `plt.scatter()` for primates and then non-primates, create a scatterplot with `cortex_mass` on the x axis and `neurons` on the y axis.\n",
    "6. Label your axes.\n",
    "7. (Optional) log scale your axes using `plt.xscale('log')` and `plt.yscale('log')`.\n",
    "8. Right click on the image or screenshot to save and submit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjoAAAG1CAYAAADwRl5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAA9hAAAPYQGoP6dpAABBOklEQVR4nO3deXiU5dn38d8kkAQIiYYlC4SwSXEMiAlBAVFxIYgmYKsolQB1xVKR4oK+VilaBW1RfCRwqE9dUQsVRbBtJK2oqUEiWwXDA4KBUBl2TQANyMz9/jFmZMg2mcx+fz/HkQPmuq+ZOcNNmJNrOy2GYRgCAACIQFHBDgAAAMBfSHQAAEDEItEBAAARi0QHAABELBIdAAAQsUh0AABAxCLRAQAAEYtEBwAARKxWwQ4g2BwOh/bs2aP27dvLYrEEOxwAAOABwzB05MgRpaWlKSqq4XEb0yc6e/bsUXp6erDDAAAAXti9e7e6du3a4HXTJzrt27eX5PyDSkhICHI0AADAE9XV1UpPT3d9jjfE9IlO7XRVQkICiQ4AAGGmqWUnLEYGAAARi0QHAABELNNPXXnC4XDoxIkTwQ4DQRYTE9Poyn4AQOgh0WnCiRMnVFFRIYfDEexQEGRRUVHq0aOHYmJigh0KAMBDJDqNMAxDNptN0dHRSk9P53/zJlZ73pLNZlO3bt04cwkAwgSJTiNOnjyp7777TmlpaWrbtm2ww0GQderUSXv27NHJkyfVunXrYIcDAPAAQxSNsNvtksRUBST99Peg9u8FACD0keh4gGkKSPw9AIBwxNQVAAB+YHcYKqs4rP1HatS5fZwG9UhSdBT/YQo0Eh0AAHysaLNNs1aUy1ZV42pLTYzTzDyrRmamBjEy82HqCpKkSy65RNOmTQt2GAAQ9oo223THovVuSY4k7a2q0R2L1qtosy1IkZmTaROdwsJCWa1W5eTk+P297A5Dq3cc0rsbv9bqHYdkdxh+fb9JkybJYrHIYrGodevW6tmzp+655x4dO3aswee8/fbbevTRR/0al0RCBSCy2R2GZq0oV33/yte2zVpR7vfPAfzEtFNXU6ZM0ZQpU1RdXa3ExES/vU+whi9Hjhypl156ST/88INKSkp0yy236NixY1q4cKFbvx9++EGtW7dWUlKS32IBALMoqzhcZyTnVIYkW1WNyioOa3CvDoELzMRMO6ITCMEcvoyNjVVKSorS09P1y1/+UjfeeKOWLVum3//+9xowYIBefPFF9ezZU7GxsTIMo85IS/fu3fWHP/xBEyZMUHx8vDIyMvTuu+/qwIEDGj16tOLj49WvXz+tXbvW9ZxDhw5p3Lhx6tq1q9q2bat+/frpzTffdF2fNGmSPvroIz3zzDOuEaedO3dKksrLyzVq1CjFx8crOTlZBQUFOnjwoN/+fADAH/YfaTjJ8aYfWo5Ex09CbfiyTZs2+uGHHyRJ27dv15IlS7R06VJt3Lixwec8/fTTGjp0qDZs2KCrrrpKBQUFmjBhgsaPH6/169erd+/emjBhggzD+T3U1NQoOztb7733njZv3qzbbrtNBQUFWrNmjSTpmWee0eDBg3XrrbfKZrPJZrMpPT1dNptNF198sQYMGKC1a9eqqKhI+/bt09ixY/3+5wIAvtS5fZxP+6HlTDt15W+hNHxZVlamN954Q5dddpkkZ/2u1157TZ06dWr0eaNGjdLtt98uSXr44Ye1cOFC5eTk6LrrrpMkzZgxQ4MHD9a+ffuUkpKiLl266J577nE9/84771RRUZH++te/6vzzz1diYqJiYmLUtm1bpaSkuPotXLhQWVlZevzxx11tL774otLT07Vt2zb16dPHZ38WAOBPg3okKTUxTnuraur9j65FUkqic6s5AoMRHT8J9vDle++9p/j4eMXFxWnw4MG66KKL9Oyzz0qSMjIymkxyJKl///6u3ycnJ0uS+vXrV6dt//79kpwnBj/22GPq37+/OnTooPj4eK1cuVKVlZWNvs+6deu0atUqxcfHu7769u0rSdqxY0czvmsACK7oKItm5lklOZOaU9U+npln5TydAGJEx0+CPXw5fPhwLVy4UK1bt1ZaWppbbaZ27dp59BqnPqf2VOD62moru8+dO1dPP/205s2bp379+qldu3aaNm2aTpw40ej7OBwO5eXl6YknnqhzLTWV8yYAhJeRmalaOD6rzkaUFM7RCQoSHT8J9vBlu3bt1Lt3b7+8dkNKSko0evRojR8/XpIzgfnyyy919tlnu/rExMTUqRWVlZWlpUuXqnv37mrVir+SAMLfyMxUXWFN4WTkEMDUlZ+Ycfiyd+/eKi4uVmlpqbZs2aLbb79de/fudevTvXt3rVmzRjt37tTBgwflcDg0ZcoUHT58WOPGjVNZWZm++uorrVy5UjfddBMFNAGEregoiwb36qDRA7pocK8OEfXvfTgh0fGj2uHLlET36amUxDgtHJ8VccOXDz30kLKyspSbm6tLLrlEKSkpGjNmjFufe+65R9HR0bJarerUqZMqKyuVlpamTz75RHa7Xbm5ucrMzNRdd92lxMRERUXxVxQA4D2LUbs32KRqDwysqqpSQkKC27WamhpVVFSoR48eiovzfi0Nhd0ig6/+PgAAWq6xz+9TsSAiAGqHLwEAQGAxLwAAACIWiQ4AAIhYJDoAACBikegAAICIRaIDAAAiFokOAACIWCQ6AAAgYpHoIKgsFouWLVsW7DAAABGKRCcCTZo0SRaLRXPmzHFrX7ZsmaviuKe6d++uefPmedTPYrHIYrGobdu2yszM1HPPPdfk82w2m6688spmxeQNEioAMCcSnUBw2KWKEmnTW85fHf4vVBkXF6cnnnhC33zzjd/fq9Yjjzwim82mzz//XGPGjNHkyZO1ePHievueOHFCkpSSkqLY2NiAxQgAMBfTJjqFhYWyWq3Kycnx7xuVL5fmZUqvXC0tvdn567xMZ7sfXX755UpJSdHs2bMb7bd06VKdc845io2NVffu3TV37lzXtUsuuUS7du3Sb3/7W9doTWPat2+vlJQU9e7dW3/4wx901llnuUZRLrnkEv3mN7/R9OnT1bFjR11xxRWS3Edadu7cKYvFoiVLlmjYsGFq06aNcnJytG3bNn322WcaOHCg4uPjNXLkSB04cMD1vp999pmuuOIKdezYUYmJibr44ou1fv161/Xu3btLkq655hpZLBbXY0lasWKFsrOzFRcXp549e2rWrFk6efJkU3+8AIAwYdpEZ8qUKSovL9dnn33mvzcpXy4tmSBV73Fvr7Y52/2Y7ERHR+vxxx/Xs88+q//+97/19lm3bp3Gjh2rG264QZs2bdLvf/97PfTQQ3r55ZclSW+//ba6du3qGqmx2WzNiiEuLk4//PCD6/Err7yiVq1a6ZNPPml0WmvmzJn63e9+p/Xr16tVq1YaN26c7rvvPj3zzDMqKSnRjh079PDDD7v6HzlyRBMnTlRJSYk+/fRTnXXWWRo1apSOHDkiSa57/NJLL8lms7kev//++xo/frymTp2q8vJyPffcc3r55Zf12GOPNev7BACELop6+ovDLhXNkFRfcXhDkkUqul/qe5UUFe2XEK655hoNGDBAM2fO1J///Oc615966ilddtlleuihhyRJffr0UXl5uf74xz9q0qRJSkpKUnR0tGukxlMnT57UokWLtGnTJt1xxx2u9t69e+vJJ59s8vn33HOPcnNzJUl33XWXxo0bp3/9618aOnSoJOnmm292JWOSdOmll7o9/7nnntOZZ56pjz76SFdffbU6deokSTrjjDPcvo/HHntM999/vyZOnChJ6tmzpx599FHdd999mjlzpsffLwAgdJl2RMfvdpXWHclxY0jVXzv7+dETTzyhV155ReXl5XWubdmyxZU81Bo6dKi+/PJL2e3NX0c0Y8YMxcfHq02bNpoyZYruvfde3X777a7rAwcO9Oh1+vfv7/p9cnKyJKlfv35ubfv373c93r9/vyZPnqw+ffooMTFRiYmJOnr0qCorKxt9n3Xr1umRRx5RfHy86+vWW2+VzWbTd99951GsAIAGBGF9an0Y0fGXo/t8289LF110kXJzc/X//t//06RJk9yuGYZRZ92NYdQ3AuWZe++9V5MmTVLbtm2Vmppa57XbtWvn0eu0bt3a9fva1zi9zeFwuB5PmjRJBw4c0Lx585SRkaHY2FgNHjzYteC5IQ6HQ7NmzdLPf/7zOtfi4uI8ihVAeLA7DJVVHNb+IzXq3D5Og3okKTqqebtQ0Qzly52zGqf+hz8hTRr5hGTND2goJDr+Ep/s234tMGfOHA0YMEB9+vRxa7darfr3v//t1lZaWqo+ffooOto5nRYTE+Px6E7Hjh3Vu3dv3wTdDCUlJVqwYIFGjRolSdq9e7cOHjzo1qd169Z1vo+srCxt3bo1KDEDaJqvkpOizTbNWlEuW1WNqy01MU4z86wamZnqy5Ah/bQ+9fSlG7XrU8e+GtBkh0THXzKGOLPXapvqX6djcV7PGOL3UPr166cbb7xRzz77rFv73XffrZycHD366KO6/vrrtXr1as2fP18LFixw9enevbs+/vhj3XDDDYqNjVXHjh39Hm9z9e7dW6+99poGDhyo6upq3XvvvWrTpo1bn+7du7vW+cTGxurMM8/Uww8/rKuvvlrp6em67rrrFBUVpc8//1ybNm3SH/7whyB9NwAk3yUnRZttumPR+jr/Cu+tqtEdi9Zr4fgskh1fCoH1qadjjY6/REU7h+gkSaf/D+THxyPnBOxGP/roo3WmpbKysrRkyRL95S9/UWZmph5++GE98sgjblNcjzzyiHbu3KlevXq5FvWGmhdffFHffPONzjvvPBUUFGjq1Knq3LmzW5+5c+equLhY6enpOu+88yRJubm5eu+991RcXKycnBxdcMEFeuqpp5SRkRGMbwPAj2qTk1OTHOmn5KRos2c7QO0OQ7NWlDf4kStJs1aUy+7wfsoepwmR9amnshgtWZQRAaqrq5WYmKiqqiolJCS4XaupqVFFRYV69Ojh/ZqNeucpuziTnADPU6JlfPL3AUCj7A5DFz7xQZ0kp5ZFUkpinP4949Imp7FW7zikcS982uR7vnnrBRrcq4M34eJ0m95ynhnXlF/8Wep3bYveqrHP71MxdeVv1nznEN2uUufC4/hk53RVgEZyACCclFUcbjDJkZwjMbaqGpVVHG4yOdl/pOHX8aYfPBBC61NrkegEQlS01GNYsKMAgJDny+Skc3vPRl497QcPhND61Fqs0QEAhAxfJieDeiQpNTGuzirJWhY5FzgP6pHkeYBoXIitT5VIdAAAIcSXyUl0lEUz86yu553+OpI0M8/KeTq+Zs13biFPOG03W0JawLeWS0xdAQBCSG1ycsei9bLIffLDm+RkZGaqFo7PqrNVPYVzdPwrhNankuh4wOQb0/Aj/h4AgeHr5GRkZqqusKZwMnKghcj6VBKdRtSeDnzixIk6B9DBfGpLStT+vQDgP75OTqKjLGwhNykSnUa0atVKbdu21YEDB9S6dWtFRbGkyawcDocOHDigtm3bqlUrfmyAQCA5gS/wL3YjLBaLUlNTVVFRoV27dgU7HARZVFSUunXrVqdYKRDqKGgJMyPRaUJMTIzOOuusJithI/LFxMQwqoewQ0FLmB2JjgeioqI48h9A2KGgJcA5OgAQkShoCTiR6ABABGpOzSggkpHoAEAEoqAl4ESiAwARiIKWgBOJDgBEIApaAk4kOgAQgShoCTiR6ABAhKqtGZWS6D49lZIYF5St5XaHodU7DundjV9r9Y5D7PhCQHCODgBEsFApaMnBhQgWi2HykszV1dVKTExUVVWVEhISgh0OAESchg4urE21OLgQ3vD085upKwCA33BwIYKNRAcA4DccXIhgI9EBAPgNBxci2CIi0fnTn/6kc845R5mZmVq0aFGwwwEA/IiDCxFsYZ/obNq0SW+88YbWrVuntWvXauHChfr222+DHRYAQBxciOAL+0Rny5YtGjJkiOLi4hQXF6cBAwaoqKgo2GEBAMTBhQi+oCc6H3/8sfLy8pSWliaLxaJly5bV6bNgwQL16NFDcXFxys7OVklJietaZmamVq1apW+//VbffvutPvjgA3399dcB/A4AAI0JtYMLYS5BPzDw2LFjOvfcc/WrX/1Kv/jFL+pcX7x4saZNm6YFCxZo6NCheu6553TllVeqvLxc3bp1k9Vq1dSpU3XppZcqMTFROTk5atUq6N8WAOAUoXJwIcwnpA4MtFgseueddzRmzBhX2/nnn6+srCwtXLjQ1Xb22WdrzJgxmj17dp3XuOWWW3TNNdfoqquuqvc9jh8/ruPHj7seV1dXKz09nQMDAQAIIxFxYOCJEye0bt06jRgxwq19xIgRKi0tdT3ev3+/JGnr1q0qKytTbm5ug685e/ZsJSYmur7S09P9EzwAAAi6kJ7jOXjwoOx2u5KTk93ak5OTtXfvXtfjMWPG6Ntvv1W7du300ksvNTp19cADD2j69Omux7UjOgAAIPKEdKJTy2Jxn8M1DMOt7dTRnabExsYqNjbWZ7EBAIDQFdJTVx07dlR0dLTb6I3knKo6fZQHAADgdCGd6MTExCg7O1vFxcVu7cXFxRoyZEiQogIAAOEi6FNXR48e1fbt212PKyoqtHHjRiUlJalbt26aPn26CgoKNHDgQA0ePFjPP/+8KisrNXny5CBGDQAAwkHQE521a9dq+PDhrse1C4UnTpyol19+Wddff70OHTqkRx55RDabTZmZmfr73/+ujIyMFr1vYWGhCgsLZbfbW/Q6AAAgdIXUOTrB4Ok+fABA89gdBgcEwm88/fwO+ogOACDyFG22adaKctmqalxtqYlxmplnpeQDAiqkFyMDAMJP0Wab7li03i3JkaS9VTW6Y9F6FW22BSkymBGJDgDAZ+wOQ7NWlKu+NRG1bbNWlMvuMPWqCf9z2KWKEmnTW85fHeZdj8rUFQDAZ8oqDtcZyTmVIclWVaOyisMa3KtD4AIzk/LlUtEMqXrPT20JadLIJyRrfvDiChLTjugUFhbKarUqJycn2KEAQMTYf6ThJMebfmim8uXSkgnuSY4kVduc7eXLgxNXEJk20ZkyZYrKy8v12WefBTsUAIgYndvH+bQfmsFhd47kNDZxWHS/6aaxTJvoAAB8b1CPJKUmxqmhTeQWOXdfDeqRFMiwzGFXad2RHDeGVP21s5+JkOgAAHwmOsqimXlWSaqT7NQ+npln5Twdfzi6z7f9IgSJDgDAp0Zmpmrh+CylJLpPT6Ukxmnh+CzO0fGXeA+LXXvaL0Kw6woA4HMjM1N1hTWFk5EDKWOIc3dVtU31r9OxOK9nmKsoNokOAMAvoqMsbCEPpKho5xbyJRPknCg8Ndn5McEcOcfZz0SYugIAIFJY86Wxr0oJp00PJqQ52014jo5pR3SoXg4AiEjWfKnvVc7dVUf3OdfkZAwx3UhOLaqXU70cAICw4+nnN1NXAAAgYpHoAACAiEWiAwAAIhaJDgAAiFim3XUFAIBPOezsdApBJDoAALRU+XJn5fBTi2ompDkP8DPh2TWhxLRTV4WFhbJarcrJyQl2KACAcFa+3Hka8emVw6ttzvby5cGJC5I4R4dzdAAA3nPYpXmZdZMclx/rS03bxDSWj3GODgAA/rartJEkR5IMqfprZz8EBYkOAADeOrrPt/3gcyQ6AAB4Kz7Zt/3gcyQ6AAB4K2OIcw2OLA10sEgJXZz9EBQkOgAAeCsq2rmFXFLdZOfHxyPnsBA5iEh0AABoCWu+NPZVKSHVvT0hzdnOOTpBxYGBAAC0lDVf6nsVJyOHIBIdAAB8ISpa6jEs2FHgNKaduuJkZAAAIh8nI3MyMgAAYcfTz2+mrgDAD+wOQ2UVh7X/SI06t4/ToB5Jio5qaAsyAH8h0QEAHyvabNOsFeWyVdW42lIT4zQzz6qRmamNPBOAr5l2jQ4A+EPRZpvuWLTeLcmRpL1VNbpj0XoVbbYFKTLAnEh0AMBH7A5Ds1aUq76Fj7Vts1aUy+4w9dJIIKBIdADAR8oqDtcZyTmVIclWVaOyisOBCwowORIdAPCR/UcaTnK86Qeg5ViMDAA+0rl9nE/7wUccdk4sNjESHQDwkUE9kpSaGKe9VTX1rtOxSEpJdG41R4CUL5eKZkjVe35qS0hzFuKkBpUpMHUFAD4SHWXRzDyrpAbrWGtmnpXzdAKlfLm0ZIJ7kiNJ1TZne/ny4MSFgCLRAQAfGpmZqoXjs5SS6D49lZIYp4XjszhHJ1AcdudITmN74Irud/ZDRDPt1FVhYaEKCwtlt/OXHIBvjcxM1RXWFE5GDqZdpXVHctwYUvXXzn4U4oxopk10pkyZoilTprhqZQCAL0VHWTS4VweP+lIuwg+O7vNtP4Qt0yY6ABAKKBfhJ/HJvu2HsMUaHQAIEspF+FHGEOfuqjrLwmtZpIQuzn6IaCQ6ABAElIvws6ho5xZySQ3ugRs5h/N0TIBEBwCCgHIRAWDNl8a+KiWcNgWYkOZs5xwdU2CNDgAEAeUiAsSaL/W9ipORTYxEBwCCgHIRARQVzRZyE2PqCgCCoLZcRCNLZZVKuQigxUh0ACAIKBcBBAaJDgAEienLRTjsUkWJtOkt56+nlmNo7BrQDKzRAYAgMm25iMaqiktUHIfPWAzDMPUhDbUlIKqqqpSQkBDscAAg8tVWFa9zipClnrZTr4lt4XDx9PObqSsAQOB4UlW8XlQch3dIdAAAgdNkVfHGnFJxHPAQiQ4AIHB8US2ciuNoBtMmOoWFhbJarcrJyQl2KABgHr6oFk7FcTQDi5FZjAwAgeOwS/MypWqbGl+TUx+Lc/fVtE2UcACLkQEAIciTquKNXaPiOJqJRAcAEFiNVhV/zflFxXH4CAcGAgACr6mq4lQch4+Q6AAAgqOxquJUHIePkOgAiBh2h2G+UgoAGkWiAyAiFG22adaKctmqalxtqYlxmplnjfzimAAaxGJkAGGvaLNNdyxa75bkSNLeqhrdsWi9ijbbghQZgGAj0QEQ1uwOQ7NWlDdaOWnWinLZHaY+MgwwLaauAIS1sorDdUZyTmVIslXV6NMdhxQVZWH9DmAyJDoAwtr+Iw0nOaea8sZ6ffv9D67HrN8BzIGpKwBhrXP7OI/6nZrkSKzfAcyCRAdAWBvUI0mpiXF1CgY0hfU7gDmQ6AAIa9FRFs3Ms0qqWx2pKbXrd8oqDvs8LgChgUQHQNgbmZmqheOzlJLoPo11RpvWHj3f03U+AMIPi5EBRISRmam6wpridjKywzB04/+uafK5nq7zARB+SHQARIzoKIsG9+rgemx3GEpNjNPeqpp6z9mxSEpJdG41BxCZmLoCELEaW79T+3hmntX35+k47FJFibTpLeevDrtvXx+AxxjRARDRatfvnF4HK8Vf5+iUL5eKZkjVe35qS0iTRj4hWfN9+14AmmQxDKPZ+yqLiooUHx+vCy+8UJJUWFioF154QVarVYWFhTrzzDN9Hqi/VFdXKzExUVVVVUpISAh2OAD8JCCVzcuXS0smSHUmyn58n7GvkuwAPuLp57dXU1f33nuvqqurJUmbNm3S3XffrVGjRumrr77S9OnTvYsYAPyodv3O6AFdNLhXB/9MVxXNUN0kRz+1Fd3PNBYQYF5NXVVUVMhqdc57L126VFdffbUef/xxrV+/XqNGjfJpgP5SWFiowsJC2e38owPAB3aVuk9X1WFI1V87+/UYFrCwALPzakQnJiZG3333nSTpn//8p0aMGCFJSkpKco30hLopU6aovLxcn332WbBDARAJju7zbT8APuHViM6FF16o6dOna+jQoSorK9PixYslSdu2bVPXrl19GiAAhIX4ZN/2A+ATXo3ozJ8/X61atdJbb72lhQsXqkuXLpKkf/zjHxo5cqRPAwSAsJAxxLm7qsFCFBYpoYuzH4CA8WrXVSRh1xUAn3HtupLcFyWz6wrwNU8/v70+R8fhcGj79u3av3+/HA6H27WLLrrI25cFgPBlzXcmM/WeozOHJAcIAq8SnU8//VS//OUvtWvXLp0+IGSxWNjJBMC8rPlS36ucu6uO7nOuyckYIkVFBzsywJS8SnQmT56sgQMH6m9/+5tSU1Nlsfj4PAoACGdR0WwhB0KEV4nOl19+qbfeeku9e/f2dTwAAAA+49Wuq/PPP1/bt2/3dSwAAAA+5dWIzp133qm7775be/fuVb9+/dS6dWu36/379/dJcAAAAC3h1fbyqKi6A0EWi0WGYYTdYmS2lwMAEH78ur28oqLC68AAAAACxatEJyMjw9dxAAAA+JzXBwbu2LFD8+bN05YtW2SxWHT22WfrrrvuUq9evXwZHwAAgNe82nX1/vvvy2q1qqysTP3791dmZqbWrFmjc845R8XFxb6OEQAAwCteLUY+77zzlJubqzlz5ri133///Vq5cqXWr1/vswD9jcXIAACEH08/v70a0dmyZYtuvvnmOu033XSTysvLvXlJAAAAn/Mq0enUqZM2btxYp33jxo3q3LlzS2MCAADwCa8WI99666267bbb9NVXX2nIkCGyWCz697//rSeeeEJ33323r2MEAADwildrdAzD0Lx58zR37lzt2bNHkpSWlqZ7771XU6dODasin6zRAQAg/PjtwMCTJ0/q9ddf17hx4/Tb3/5WR44ckSS1b9/e+2gBAAD8oNlrdFq1aqU77rhDx48fl+RMcEhyAABAKPK6evmGDRt8HQsAAIBPebUY+de//rXuvvtu/fe//1V2drbatWvndp3q5QAAIBRQvZzFyAAAhB2qlwMAANOjejkANJfDLu0qlY7uk+KTpYwhUlR0sKMCUA+vEp1XX3210esTJkzwKhgACHnly6WiGVL1np/aEtKkkU9I1vzgxQWgXl6t0TnzzDPdHv/www/67rvvFBMTo7Zt2+rw4cM+C9DfWKMDwGPly6UlEySd/s/mj4ekjn2VZAcIEL8W9fzmm2/cvo4ePaqtW7fqwgsv1Jtvvul10AAQshx250hOnSRHP7UV3e/sByBkeJXo1Oess87SnDlzdNddd/nqJQEgdOwqdZ+uqsOQqr929gMQMnyW6EhSdHS0q/ZVID399NM655xzZLVaNXXqVHkxGwcAjTu6z7f9AASEV4uRly9f7vbYMAzZbDbNnz9fQ4cO9Ulgnjpw4IDmz5+vL774Qq1bt9ZFF12kTz/9VIMHDw5oHAAiREM7quKTPXu+p/0ABIRXic6YMWPcHlssFnXq1EmXXnqp5s6d64u4muXkyZOqqamR5FwY3blz54DHACACNLajqu9Vzt9X21T/Oh2L83rGkEBFC8ADXk1dORwOty+73a69e/fqjTfeUGpqarNe6+OPP1ZeXp7S0tJksVi0bNmyOn0WLFigHj16KC4uTtnZ2SopKXFd69Spk+655x5169ZNaWlpuvzyy9WrVy9vvi0AZla7o+r0dTjVNmf7//3NmfBIcu2ycvnx8cg5nKcDhJgWrdE5ceKEtm7dqpMnT3r9GseOHdO5556r+fPn13t98eLFmjZtmh588EFt2LBBw4YN05VXXqnKykpJzh1g7733nnbu3Kmvv/5apaWl+vjjj72OB4AJebqjqu9Vzi3kCaf9hy4hja3lQIjyaurqu+++029+8xvXwYHbtm1Tz549NXXqVKWlpen+++/3+LWuvPJKXXnllQ1ef+qpp3TzzTfrlltukSTNmzdP77//vhYuXKjZs2frn//8p3r37q2kpCRJ0lVXXaVPP/1UF110Ub2vd/z4cR0/ftz1uLq62uNYAUSo5uyosuY7Ex5ORgbCglcjOg888IA+//xzffjhh4qLi3O1X3755Vq8eLHPgjtx4oTWrVunESNGuLWPGDFCpaXOLZzp6ekqLS1VTU2N7Ha7PvzwQ/3sZz9r8DVnz56txMRE11d6errP4gUQppq7oyoqWuoxTOp3rfNXkhwgZHmV6Cxbtkzz58/XhRdeKIvlp7lqq9WqHTt2+Cy4gwcPym63KznZfRdDcnKy9u7dK0m64IILNGrUKJ133nnq37+/evXqpfz8hoePH3jgAVVVVbm+du/e7bN4AYQpdlQBEcurqasDBw7Uu7Pp2LFjbomPr5z+moZhuLU99thjeuyxxzx6rdjYWMXGxvo0PgBhLmMIO6qACOXViE5OTo7+9re/uR7XJh0vvPCCT8+v6dixo6Kjo12jN7X2799fZ5QHALwWFc2OKiBCeTWiM3v2bI0cOVLl5eU6efKknnnmGX3xxRdavXq1PvroI58FFxMTo+zsbBUXF+uaa65xtRcXF2v06NE+ex8AkDXfuXOq3nN05rCjCghTXiU6Q4YM0SeffKI//elP6tWrl1auXKmsrCytXr1a/fr1a9ZrHT16VNu3b3c9rqio0MaNG5WUlKRu3bpp+vTpKigo0MCBAzV48GA9//zzqqys1OTJk70JHQAaxo4qIOJYjCAXhvrwww81fPjwOu0TJ07Uyy+/LMl5YOCTTz4pm82mzMxMPf300w1uH/dUYWGhCgsLZbfbtW3btibLvAMAgNBRXV2txMTEJj+/m5XoREVFNbnY2GKxtOgAwUDz9A8KAACEDk8/v5s1dfXOO+80eK20tFTPPvsslcMBAEDIaFaiU98C4P/7v//TAw88oBUrVujGG2/Uo48+6rPgAISwhqp8A0AI8WoxsiTt2bNHM2fO1CuvvKLc3Fxt3LhRmZmZvowNQKhqrMo3u5MAhJBmn6NTVVWlGTNmqHfv3vriiy/0r3/9SytWrCDJAcyiqSrf5cuDExcA1KNZic6TTz6pnj176r333tObb76p0tJSDRs2zF+x+VVhYaGsVqtycnKCHQoQPjyt8u2wBzIqAGhQs3ddtWnTRpdffrmioxuei3/77bd9ElwgsOsKaIaKEumVq5vuN/E9Z7FLX2AtEIB6+GXX1YQJE/xSywpAmGhule+WYi0QgBZqVqJTe4AfAJMKZJXv2rVAp0+T1a4FGvsqyQ6AJnlV1BOASdVW+a5T+LKWRUro0vIq36wFAuAjJDoAPBeoKt+7Suvu6nJjSNVfO/sBQCNIdAA0T22V74RU9/aENN9NJwV6LRCAiOX1gYHh7tSingCayd9VvgO5FghARAt69fJgY3s5EIIcdmlepnPhcb3rdCzOEaRpm9hqDpiUp5/fTF0BCD2BWgsEIOKR6ADhxmF3Hty36S3nr5G68ygQa4EARDzTrtEBwpLZDtDz91ogABGPRAcIF2Y9QC8q2nflJACYDlNXQDjgAD0A8AqJDhAOOEAPALxCogOEAw7QAwCvmDbRKSwslNVqVU5OTrBDAZrGAXoA4BUODOTAQISDUDlAz2FnBxSAkODp5ze7roBwUHuA3pIJch6Yd2qyE6AD9My2tR1ARDDt1BUQdoJ5gF7t1vbTF0TXbm0vX+6/9waAFmBEBwgnwThAr8mt7Rbn1va+VzGNBSDkkOgA4SbQB+g1Z2s7B/sBCDFMXQFoHFvbAYQxEh0AjWNrO4AwRqIDoHEZQ5wLnmt3d9VhkRK6OPsBQIgh0QHQuNqt7ZLqJjsB2toOAF4ybaLDychAMwRzazsAtAAnI3MyMuA5TkYGECI4GRlAXS1NVAK9tR0AWohEBzALSjgAMCHTrtEBTIUSDgBMikQHiHRNlnCQs4SDwx7IqAAgIEh0gEjXnBIOABBhSHSASEcJBwAmRqIDRDpKOAAwMRIdINJRwgGAiZHoAJGOEg4ATIxEBzADSjgAMCnTHhhYWFiowsJC2e1sqYVJWPOlvldRwgGAqVDrilpXAACEHU8/v5m6AgAAEYtEBwAARCzTrtEBQlJLq4sDANyQ6AChguriAOBzTF0BoYDq4gDgFyQ6QLBRXRwA/IZEBwg2qosDgN+wRgfwlL8WClNdHAD8hkQH8IQ/FwpTXRwA/IapK6Ap/l4oTHVxAPAbEh2gMYFYKEx1cQDwGxIdoDGBWihMdXEA8AvW6ACNCeRCYaqLA4DPkegAjQn0QuGoaKnHMN+8FgDAvFNXhYWFslqtysnJCXYoCGUsFAaAsGYxDKO+VZamUV1drcTERFVVVSkhISHY4cBb/iyGWbvrSpL7ouQfkx/W0ABAwHn6+c3UFcKfv4th1i4Urvc95tR9DyqQA0DIYESHEZ3wVr5cWlJQzwU/jLZ4ksBQgRwAAsLTz28SHRKd8OWwS3/sJX3/TQMdLM4kY9qmwIyouKa4Tv+RYooLAHzN089v0y5GRgT4+E+NJDlSQIthUoEcAEISiQ7Ck8MurVnoWd9AFMOkAjkAhCQSHYSnXaVNjOacIhDFMKlADgAhiUQH4cnThKHNmYE544YK5AAQkkh0EJ48TRjOvyMwC5E5WBAAQhKJDsJTk4mFpDZJ0kX3BCYeKpADQEgi0UF4ajSx+LEt75nAJhbWfNmve0XH25422kQFcgAIGk5GRvhq8MTiLvWfWOxnRZttmrUiXvuq/qRBUf+nzvpWJ9t1Vv4Vv9BIa9eAxgIAcOLAQA4MDD+nn1Ccfr60e01QSy4UbbbpjkXrGzoqUAvHZ2lkZmpAYwKASEatK0Smxkos9Ls2KCHZHYZmrShv8KhAi6RZK8p1hTVF0VGNrCkCAPgca3QQPmpLLJx+MF+1zdlevjwoYZVVHJatqqbB64YkW1WNyioOBy4oAIAkEh2EixAusbD/SMNJjjf9AAC+Q6KD8BDCJRY6t4/zaT8AgO+Q6CA8hHCJhUE9kpSaGNfYUYFKTYzToB5JgQwLACASHYSLEC6xEB1l0cw8q6QGjwrUzDwrC5EBIAhIdBAeQrzEwsjMVC0cn6WURPfpqZTEOLaWA0AQsb0c4aH2JOQlE+RMdk5dlBwaJRZGZqbqCmuKyioOa/+RGnVu75yuYiQHAILHtIlOYWGhCgsLZbcHfpcOvNTgSchpQTkJuT7RURYN7tUh2GEAAH7EycicjBx+Tj8ZOQgnIQMAgouTkRG5oqKlHsOCHQUAIAyQ6CB8nDqS07ajZLFIxw4wqgMAaBCJDsJDfTWuTlVb7yoE1ukAAEIH28sR+hqqcXWqINe7AgCEJhIdhLZGa1ydKrj1rgAAoYlEB6GtyRpXpwpevSsAQGgi0UFo86Z2VRDqXQEAQhOJDkKbN7WrglDvCgAQmkh0ENpcNa48Edx6VwCA0EOig9AWFS2NmO15/yDXuwIAhBYSHYS+dh7WjrrkAc7RAQC4IdFB6PN0cXGHXv6NAwAQdkh0EPo8XVzMImQAwGlIdBD6XAuSLQ10YBEyAKB+JDpm47BLFSXSprecv4bDKcJR0c46VpLqJjs/PmYRMgCgHhT1NJP6CmOGSzFMa7409tUG4p8T+vEDAILCYhhGU0WEIlp1dbUSExNVVVWlhISEYIfjP7WFMevUjPpxRGTsq+GRLDjszhIPR/c51+RkDGEkBwBMyNPPb0Z0zKDRwpiGJIuzGGbfq0I/aYiKlnoMC3YUAIAwwRodM2iyMCbFMAEAkYlExww8PYeGYpgAgAhDomMGnEMDADAp1uiYQe05NNU21b9Ox+K8nn6+c8s5C30BABGCRMcMas+hWTJBzl1WpyY7P+66yvyF9D/nhufWcwAAGsDUlVnUnkOTkOrenpAmDblTKn227oLlapszOSpfHrg4AQDwIUZ0zMSa79xCfuo5NOnnO0dyImHrOQAApyHRMZvTz6GpKPF86znn1wAAwgxTV2bH1nMAQAQj0TE7tp4DACIYiY7Z1W49r1MVvJZFSuji7AcAQJgh0TG72q3nkuomOz8+HjmHhcgAgLAU9onO1q1bNWDAANdXmzZttGzZsmCHFV4a23oeLlXNAQCoh8UwjPr2FYelo0ePqnv37tq1a5fatWvn0XM8LfNuCg67+9ZzTkYGAIQoTz+/I2p7+fLly3XZZZd5nOT4TbgmDKdvPQcAIMwFferq448/Vl5entLS0mSxWOqddlqwYIF69OihuLg4ZWdnq6SkpN7XWrJkia6//no/R9yE8uXSvEzplaulpTc7f52XyenCAAAEQdATnWPHjuncc8/V/Pnz672+ePFiTZs2TQ8++KA2bNigYcOG6corr1RlZaVbv+rqan3yyScaNWpUIMKuX/lyZ8kESikAABASQmqNjsVi0TvvvKMxY8a42s4//3xlZWVp4cKFrrazzz5bY8aM0ezZs11tr732mt5//30tWrSo0fc4fvy4jh8/7npcXV2t9PT0lq/RcdidIzcNnjL8Y4XwaZvCYxoLAIAQ5ukanaCP6DTmxIkTWrdunUaMGOHWPmLECJWWlrq1eTptNXv2bCUmJrq+0tPTfRPsrlLPSykAAICACOlE5+DBg7Lb7UpOdj+VNzk5WXv37nU9rqqqUllZmXJzc5t8zQceeEBVVVWur927d/smWEopAAAQcsJi15XF4n6QnWEYbm2JiYnat8+zBCI2NlaxsbE+jU8SpRQAAAhBIT2i07FjR0VHR7uN3kjS/v3764zyBB2lFAAACDkhnejExMQoOztbxcXFbu3FxcUaMiTEEgZKKQAAEHKCPnV19OhRbd++3fW4oqJCGzduVFJSkrp166bp06eroKBAAwcO1ODBg/X888+rsrJSkydPDmLUDagtpVA0w31hckKaM8mhlAIAAAEV9ERn7dq1Gj58uOvx9OnTJUkTJ07Uyy+/rOuvv16HDh3SI488IpvNpszMTP39739XRkZGi963sLBQhYWFstvtLXqdOqz5Ut+rwvNkZAAAIkxInaMTDNS6AgAg/ETEOToAAAAtQaIDAAAiFokOAACIWCQ6AAAgYpk20SksLJTValVOTk6wQwEAAH7Crit2XQEAEHbYdQUAAEyPRAcAAEQsEh0AABCxSHQAAEDEItEBAAARy7SJDtvLAQCIfGwvZ3s5AABhx9PP71YBjAmhxmGXdpVKR/dJ8clSxhApKjrYUQEA4DMkOmZVvlwqmiFV7/mpLSFNGvmEZM0PXlwAAPiQadfomFr5cmnJBPckR5Kqbc728uXBiQsAAB8j0TEbh905kqP6lmb92FZ0v7MfAABhjkTHbHaV1h3JcWNI1V87+wEAEOZIdMzm6D7f9gMAIISR6JhNfLJv+wEAEMJMm+iY9sDAjCHO3VWyNNDBIiV0cfYDACDMcWCgGQ8MrN11Jcl9UfKPyc/YV9liDgAIaZ5+fpt2RMfUrPnOZCYh1b09IY0kBwAQUTgw0Kys+VLfqzgZGQAQ0Uh0zCwqWuoxLNhRAADgN0xdAQCAiEWiAwAAIhaJDgAAiFgkOgAAIGKR6AAAgIhl2kTHtCcjAwBgIpyMbMaTkQEACHOcjAwAAEyPRAcAAEQs05+MXDtzV11dHeRIAACAp2o/t5tagWP6ROfIkSOSpPT09CBHAgAAmuvIkSNKTExs8LrpFyM7HA7t2bNH7du3l8VikSTl5OTos88+a/A53l4/vb26ulrp6enavXt3SC2Ebur7C8brNve5nvZvyb329D5L3Gt/Ppd77Rl/3OuWvmZznh/M+1zfNTPd55a+rr9+pgcOHKgPPvhAaWlpiopqeCWO6Ud0oqKi1LVrV7e26OjoRv/ienu9ofaEhISQ+kFp6vsLxus297me9m/JvW7ufZa41/54LvfaM/641y19zeY8P5j3ubFrZrjPLX1df/1Mt2rVqs7nd31YjFyPKVOm+OV6U88LFf6KsyWv29znetq/Jfc63O+zxL329Dr32j+v2ZznB/M+N+f9g42f6bpMP3UVTJzhYx7ca/PgXpsD9zl8MKITRLGxsZo5c6ZiY2ODHQr8jHttHtxrc+A+hw9GdAAAQMRiRAcAAEQsEh0AABCxSHQAAEDEItEBAAARi0QHAABELBKdMHLNNdfozDPP1LXXXhvsUOBD7733nn72s5/prLPO0v/+7/8GOxz4ET/D5rB7925dcsklslqt6t+/v/76178GOyRTY3t5GFm1apWOHj2qV155RW+99Vaww4EPnDx5UlarVatWrVJCQoKysrK0Zs0aJSUlBTs0+AE/w+Zgs9m0b98+DRgwQPv371dWVpa2bt2qdu3aBTs0U2JEJ4wMHz5c7du3D3YY8KGysjKdc8456tKli9q3b69Ro0bp/fffD3ZY8BN+hs0hNTVVAwYMkCR17txZSUlJOnz4cHCDMjESHR/5+OOPlZeXp7S0NFksFi1btqxOnwULFqhHjx6Ki4tTdna2SkpKAh8ofKql933Pnj3q0qWL63HXrl319ddfByJ0NBM/4+bhy3u9du1aORwOpaen+zlqNIREx0eOHTumc889V/Pnz6/3+uLFizVt2jQ9+OCD2rBhg4YNG6Yrr7xSlZWVrj7Z2dnKzMys87Vnz55AfRtoppbe9/pmji0Wi19jhnd88TOO8OCre33o0CFNmDBBzz//fCDCRkMM+Jwk45133nFrGzRokDF58mS3tr59+xr3339/s1571apVxi9+8YuWhgg/8Oa+f/LJJ8aYMWNc16ZOnWq8/vrrfo8VLdOSn3F+hsOLt/e6pqbGGDZsmPHqq68GIkw0ghGdADhx4oTWrVunESNGuLWPGDFCpaWlQYoK/ubJfR80aJA2b96sr7/+WkeOHNHf//535ebmBiNctAA/4+bhyb02DEOTJk3SpZdeqoKCgmCEiVO0CnYAZnDw4EHZ7XYlJye7tScnJ2vv3r0ev05ubq7Wr1+vY8eOqWvXrnrnnXeUk5Pj63DhI57c91atWmnu3LkaPny4HA6H7rvvPnXo0CEY4aIFPP0Z52c4/Hlyrz/55BMtXrxY/fv3d63vee2119SvX79AhwuR6ATU6WsvDMNo1noMduOEp6bue35+vvLz8wMdFvygqXvNz3DkaOxeX3jhhXI4HMEIC/Vg6ioAOnbsqOjo6DqjN/v376/zvwJEDu67eXCvzYN7HX5IdAIgJiZG2dnZKi4udmsvLi7WkCFDghQV/I37bh7ca/PgXocfpq585OjRo9q+fbvrcUVFhTZu3KikpCR169ZN06dPV0FBgQYOHKjBgwfr+eefV2VlpSZPnhzEqNFS3Hfz4F6bB/c6wgR1z1cEWbVqlSGpztfEiRNdfQoLC42MjAwjJibGyMrKMj766KPgBQyf4L6bB/faPLjXkYVaVwAAIGKxRgcAAEQsEh0AABCxSHQAAEDEItEBAAARi0QHAABELBIdAAAQsUh0AABAxCLRAQAAEYtEBwAC7M9//rNGjBjRrOdce+21euqpp/wUERC5SHQAuOzdu1d33nmnevbsqdjYWKWnpysvL0//+te/Wvzal1xyiaZNm9byIJvw4YcfymKx6Mwzz1RNTY3btbKyMlksFlksFr/H0ZDjx4/r4Ycf1kMPPdSs5z388MN67LHHVF1d7afIgMhEogNAkrRz505lZ2frgw8+0JNPPqlNmzapqKhIw4cP15QpU7x+3R9++MGHUXquffv2euedd9zaXnzxRXXr1i0o8dRaunSp4uPjNWzYsGY9r3///urevbtef/11P0UGRCYSHQCSpF//+teyWCwqKyvTtddeqz59+uicc87R9OnT9emnn7r6VVZWavTo0YqPj1dCQoLGjh2rffv2ua7//ve/14ABA/Tiiy+6RoYmTpyojz76SM8884xrRGXnzp2SpPLyco0aNUrx8fFKTk5WQUGBDh48KMk5OhMTE6OSkhLX68+dO1cdO3aUzWZr9PuZOHGiXnzxRdfj77//Xn/5y180ceJEt36HDh3SuHHj1LVrV7Vt21b9+vXTm2++6dbnrbfeUr9+/dSmTRt16NBBl19+uY4dO+aKcdCgQWrXrp3OOOMMDR06VLt27Wowrr/85S/Kz893azt58qSmTp2qM844Qx06dNCMGTM0ceJEjRkzxq1ffn5+ndgANI5EB4AOHz6soqIiTZkyRe3atatz/YwzzpAkGYahMWPG6PDhw/roo49UXFysHTt26Prrr3frv337di1ZskRLly7Vxo0b9T//8z8aPHiwbr31VtlsNtlsNqWnp8tms+niiy/WgAEDtHbtWhUVFWnfvn0aO3aspJ+muwoKClRVVaX//Oc/evDBB/XCCy8oNTW10e+poKBAJSUlqqyslOQcSenevbuysrLc+tXU1Cg7O1vvvfeeNm/erNtuu00FBQVas2aNJMlms2ncuHG66aabtGXLFn344Yf6+c9/LsMwdPLkSY0ZM0YXX3yxPv/8c61evVq33XZbo1NjJSUlGjhwoFvbE088oddff10vvfSSPvnkE1VXV2vZsmV1njto0CCVlZXp+PHjjX7vAE4R3OLpAELBmjVrDEnG22+/3Wi/lStXGtHR0UZlZaWr7YsvvjAkGWVlZYZhGMbMmTON1q1bG/v373d77sUXX2zcddddbm0PPfSQMWLECLe23bt3G5KMrVu3GoZhGMePHzfOO+88Y+zYscY555xj3HLLLY3GuGrVKkOS8c033xhjxowxZs2aZRiGYQwfPtx45plnjHfeecdo6p++UaNGGXfffbdhGIaxbt06Q5Kxc+fOOv0OHTpkSDI+/PDDRl+v1jfffGNIMj7++GO39uTkZOOPf/yj6/HJkyeNbt26GaNHj3br95///KfBWADUjxEdADIMQ5KaXKS7ZcsWpaenKz093dVmtVp1xhlnaMuWLa62jIwMderUqcn3XbdunVatWqX4+HjXV9++fSVJO3bskCTFxMRo0aJFWrp0qb7//nvNmzfP4+/rpptu0ssvv6yvvvpKq1ev1o033linj91u12OPPab+/furQ4cOio+P18qVK10jQeeee64uu+wy9evXT9ddd51eeOEFffPNN5KkpKQkTZo0Sbm5ucrLy9MzzzzT6JTa999/L0mKi4tztVVVVWnfvn0aNGiQqy06OlrZ2dl1nt+mTRtJ0nfffefxnwFgdiQ6AHTWWWfJYrG4JSv1MQyj3mTo9Pb6pr/q43A4lJeXp40bN7p9ffnll7roootc/UpLSyU5p9gOHz7s0WtL0qhRo1RTU6Obb75ZeXl56tChQ50+c+fO1dNPP6377rtPH3zwgTZu3Kjc3FydOHFCkjPpKC4u1j/+8Q9ZrVY9++yz+tnPfqaKigpJ0ksvvaTVq1dryJAhWrx4sfr06eO2pulUHTp0kMVicSVKpzr9z7U2+TxV7ffuSRIJwIlEB4CSkpKUm5urwsJC1yLbU3377beSnKM3lZWV2r17t+taeXm5qqqqdPbZZzf6HjExMbLb7W5tWVlZ+uKLL9S9e3f17t3b7as2WdqxY4d++9vf6oUXXtAFF1ygCRMmyOFwePR9RUdHq6CgQB9++KFuuummevuUlJRo9OjRGj9+vM4991z17NlTX375pVsfi8WioUOHatasWdqwYYNiYmLcdnSdd955euCBB1RaWqrMzEy98cYbDf4ZWK1WlZeXu9oSExOVnJyssrIyV5vdbteGDRvqPH/z5s3q2rWrOnbs6NH3D4BEB8CPFixYILvdrkGDBmnp0qX68ssvtWXLFtdCYkm6/PLL1b9/f914441av369ysrKNGHCBF188cV1Ftiernv37lqzZo127typgwcPyuFwaMqUKTp8+LDGjRunsrIyffXVV1q5cqVuuukm2e122e12FRQUaMSIEfrVr36ll156SZs3b9bcuXM9/r4effRRHThwQLm5ufVe7927t4qLi1VaWqotW7bo9ttv1969e13X16xZo8cff1xr165VZWWl3n77bR04cEBnn322Kioq9MADD2j16tXatWuXVq5cqW3btjWa9OXm5urf//63W9udd96p2bNn691339XWrVt111136ZtvvqkzylNSUtLsgwYB0wvuEiEAoWTPnj3GlClTjIyMDCMmJsbo0qWLkZ+fb6xatcrVZ9euXUZ+fr7Rrl07o3379sZ1111n7N2713V95syZxrnnnlvntbdu3WpccMEFRps2bQxJRkVFhWEYhrFt2zbjmmuuMc444wyjTZs2Rt++fY1p06YZDofDmDVrlpGammocPHjQ9TrLli0zYmJijA0bNtT7PZy6GLk+py9GPnTokDF69GgjPj7e6Ny5s/G73/3OmDBhgmshcHl5uZGbm2t06tTJiI2NNfr06WM8++yzhmEYxt69e40xY8YYqampRkxMjJGRkWE8/PDDht1ub/DPeMuWLUabNm2Mb7/91tX2ww8/GL/5zW+MhIQE48wzzzRmzJhhXHfddcYNN9zg6vP9998bCQkJxurVqxt8bQB1WQyjnolgAIDfjB071jXdVR+Hw6Gzzz5bY8eO1aOPPipJKiws1LvvvquVK1cGMlQg7DF1BQAB9sc//lHx8fGux7t27dILL7ygbdu2adOmTbrjjjtUUVGhX/7yl64+rVu31rPPPhuMcIGwxogOAATZ7t27dcMNN2jz5s0yDEOZmZmaM2eO284zAN4h0QEAABGLqSsAABCxSHQAAEDEItEBAAARi0QHAABELBIdAAAQsUh0AABAxCLRAQAAEYtEBwAARCwSHQAAELH+P4vkTdQ1PyDMAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# example of a double scatter\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "brains_df = pd.read_csv('Data/species_brainmass_neurons.csv')\n",
    "\n",
    "primate_df = brains_df[brains_df['Order'] == 'Primata']\n",
    "nonprimate_df = brains_df[brains_df['Order'] != 'Primata']\n",
    "\n",
    "plt.scatter(primate_df['cortex_mass_g'],primate_df['Neurons'],label='Primate')\n",
    "plt.scatter(nonprimate_df['cortex_mass_g'],nonprimate_df['Neurons'],label='Not Primate')\n",
    "plt.legend()\n",
    "plt.ylabel('Neurons')\n",
    "plt.xlabel('Cortex Mass (g)')\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "## Additional resources\n",
    "See the [Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/03.00-introduction-to-pandas.html) for a more in depth exploration of Pandas, and of course, the [Pandas documentation](https://pandas.pydata.org/docs/user_guide/index.html)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "rise": {
   "theme": "serif"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
